---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

Transcription factor Occupancy Profiler (`TOP`)
=================================================

<!-- badges: start -->
<!-- badges: end -->

TOP fits a Bayesian hierarchical model 
using transcription factor (TF) motifs, DNase- or ATAC-seq data,
as well as ChIP-seq data (only required in training) 
from multiple TFs across multiple cell types. 

It can be used to predict the quantitative occupancy or binding probability
for many TFs using data from a single DNase- or ATAC-seq experiment. 
Thus, it allows efficient profiling of quantitative TF occupancy landscapes 
across multiple cell types or conditions using DNase- or ATAC-seq experiments.


```{r TOP-diagram, echo=FALSE, out.width = '90%'}
knitr::include_graphics("man/figures/schematic.png")
```

TOP websites
------------

  * [TOP R package website][top-webiste]
  * [TOP paper resources website][top-resources-website]: 
  a companion website for the `TOP` package, which 
  contains the data resources and data processing pipelines for the `TOP` paper.


Install TOP R package
---------------------
You can install the development version of `TOP`
from [GitHub][github] with:

``` r
# install.packages("devtools")
devtools::install_github("HarteminkLab/TOP")
```

After installing, check that it loads properly:
``` r
library(TOP)
```

Tutorials
---------
Please follow the [tutorials][tutorials] to learn how to use the package. 

Reference
----------------------
* Luo K, Zhong J, Safi A, Hong LK, Tewari AK, Song L, Reddy TE, Ma L, 
Crawford GE, Hartemink AJ. 
Profiling the quantitative occupancy of myriad transcription factors across 
conditions by modeling chromatin accessibility data. 
Genome Res. 2022 Jun;32(6):1183-1198. doi: 10.1101/gr.272203.120.

License
---------
All source code and software in this repository are made available
under the terms of the [MIT license][mit-license].

[top-webiste]: https://HarteminkLab.github.io/TOP
[top-resources-website]: https://harteminklab.github.io/TOP-paper-resources
[github]: https://github.com/HarteminkLab/TOP
[mit-license]: https://opensource.org/licenses/mit-license.html
[tutorials]: https://HarteminkLab.github.io/TOP/articles/index.html


