---
title: "Process DNase-seq bam files to data matrices"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Process DNase-seq bam files to data matrices}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Step 0: Download JASPAR motif database and install required software tools

Download [JASPAR](http://jaspar.genereg.net) motif database

Required softwares:

- [FIMO](http://meme-suite.org/doc/fimo.html?man_type=web) from MEME suite.
- [samtools](http://www.htslib.org)
- [bedtools](http://bedtools.readthedocs.io/en/latest/)
- [bwtools](https://github.com/CRG-Barcelona/bwtool/wiki)
- [UCSC command-line tools](http://hgdownload.cse.ucsc.edu/admin/exe/): `bedGraphToBigWig`, `bigWigAverageOverBed` 

## Step 1: find TF motif matches using [FIMO](http://meme-suite.org/doc/fimo.html?man_type=web)

* [fimo_jaspar_motif.sh](../script/sites_jaspar_motif.sh)

## Step 2: get TF candidate binding sites

* [sites_jaspar_motif.sh](../script/sites_jaspar_motif.sh)
* [process_pwm_sites.R](../script/process_pwm_sites.R)

## Step 3: count DNase-seq genome-wide cleavage, and build tagcount bigwig file

* [genome_coverage_bamToBigwig.sh](../script/genome_coverage_bamToBigwig.sh)

## Step 4: match DNase-seq tagcount matrices for each motif

* [get_motif_count_matrices.sh](../script/get_motif_count_matrices.sh)

* [rev_tagcount_bwtool.R](../script/rev_tagcount_bwtool.R)


```{r setup}
library(TOP)
```
