---
title: "Pipeline to prepare input data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Prepare input data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

We provided scripts and a Snakemake pipeline to prepare candidate binding sites, 
DNase- or ATAC-seq bins from DNase- or ATAC-seq reads (bam files). 

## Basic procedure
You can run the following steps as described here. 
You can use the Snakemake pipeline in the `snakemake` directory, especially
you want to train your own model with data from many TFs and cell types.

### Download motif database and install required softwares

* Download motifs from [JASPAR](http://jaspar.genereg.net) motif database

* Install softwares that will be needed later
    - [FIMO](http://meme-suite.org/doc/fimo.html?man_type=web) from MEME suite.
    - [samtools](http://www.htslib.org)
    - [bedtools](http://bedtools.readthedocs.io/en/latest/)
    - [bwtools](https://github.com/CRG-Barcelona/bwtool/wiki)
    - [UCSC command-line tools](http://hgdownload.cse.ucsc.edu/admin/exe/): 
    `bedGraphToBigWig`, `bigWigAverageOverBed` 

### Prepare candidate sites

**Major steps**:

**Step 1: Find TF motif matches using FIMO software**

To scan for TF motif matches, we recommend the `FIMO` software from the MEME suite:

* FIMO instructions: http://meme-suite.org/doc/fimo.html?man_type=web

* MEME Suite installation: http://meme-suite.org/doc/install.html?man_type=web

You will need the FIMO output (fimo.txt) in the following step. 

You can follow the instructions from [FIMO](http://meme-suite.org/doc/fimo.html?man_type=web).

We use the command line version of FIMO. By default, we used 
with the threshold $p < 1e-5$ and uniform background when training the model. 
You can choose your own background or use FIMO's default threshold $p < 1e-4$ 
(which will result in more motif matches).
```{r get-motif-matches, eval = FALSE}
# Example setting
# outdir is the output directory
# pwm: PWM file
# ref_genome: reference genome file
fimo --text --skip-matched-sequence --verbosity 2 \
  --bgfile --uniform-- \
  --thresh 1e-5 --max-stored-scores 1000000 \
  --oc outdir \
  pwm ref_genome
```

**Step 2: Get candidate TF binding sites**

Get candidate TF binding sites with 100bp flanking regions around motif matches obtained from FIMO
```{r process_candidate_sites, eval = FALSE}
# fimo: FIMO result filename. (fimo.txt)
# blacklist: ENCODE blacklist file
# flank: size (bp) of flanking region on each side of motif. Default: 100.
# thresh_pValue: FIMO p-value threshold. 
Rscript ../scripts/get_candidate_sites.R \
        --fimo fimo --blacklist blacklist \
        --flank 100 --thresh_pValue 1e-5 \
        --out output
```

**Step 3: Count DNase- or ATAC-seq genome-wide cleavage**

Sort and index the BAM file, and then retrieve and print stats in the index file
```{r bam_sort_index_stats, eval = FALSE}
samtools index bam 
samtools idxstats bam > bam.idxstats
```

Count the DNase- or ATAC-seq cleavage along the genome, output in Bigwig format
```{r count_genome_coverage, eval = FALSE}
# bam: input BAM filename.
# chrom_size: filename of genome sizes by chromosomes.
# out_dir: output directory
# out_prefix: output prefix of the counts for genome-wide cuts
Rscript ../scripts/count_dnase_genome_cuts.R \
        --bam bam \
        --chrom_size chrom_size \
        --out_dir genome_counts_dir \
        --out_prefix sample_name

```

Step 2: Get DNase- or ATAC-seq count matrices for each motif's candidate sites
```{r match_tagcount_sites, eval = FALSE}
# sites: filename for candidate sites
# dnase_fwd: filename for DNase- or ATAC-seq genome cleavage counts in forward strand
# dnase_rev: filename for DNase- or ATAC-seq genome cleavage counts in reverse strand
# outdir: output directory
# outname: output prefix of the counts for the motif
Rscript ../scripts/get_dnase_motif_counts.R \
        --sites sites \
        --dnase_fwd fwd_genome_counts \
        --dnase_rev rev_genome_counts\
        --outdir outdir \
        --outname outname
```

Step 3: Normalize, bin DNase- or ATAC-seq counts with MILLIPEDE binning scheme
and transform (asinh) DNase counts
```{r normalize_bin_dnase, eval=FALSE}
# dnase_counts: DNase- or ATAC-seq cleavage matrix
# idxstats: idxstats file generated by samtools
# bin: type of MILLIPEDE binning scheme (Default: 'M5').
# ref_size: normalize to reference library size (Default: 1e8 for DNase-seq and 5e7 for ATAC-seq)
# transform: log2 or asinh transform
Rscript ../scripts/normalize_bin_dnase.R \
        --dnase_counts dnase_counts \
        --idxstats idxstats \
        --bin 'M5' \
        --ref_size 1e8 \
        --transform 'asinh' \
        --outdir {outdir} \
        --outname {outname}
```

### Prepare ChIP-seq data if you need to train your own model

Step 1: Normalize, count ChIP-seq read coverage
```{r count_chipseq_coverage, eval=FALSE}
# pwm_id: motif pwm ID.
# cell_type: cell type.
# sites: Filename for candidate sites.
# chrom_size: Filename listing the size for each chromosome.
# metadata: Filename for training metadata table.
# chip_dir: Directory for ChIP-seq bam files.
# ref_size: Reference library size for ChIP-seq. (Default: 2e7)
# outdir: Output directory.
# outname: Output filename prefix.
Rscript ../scripts/count_chipseq_coverage.R \
        --pwm_id pwm_id \
        --cell_type cell_type \
        --sites sites \
        --chip_dir chip_dir \
        --chrom_size chrom_size \
        --metadata chrom_size \
        --ref_size 2e7 \
        --outdir chip_counts_dir \
        --outname outname
```

Step 2: Combine PWM scores, DNase (or ATAC) bins, and ChIP-seq counts for candidate sites
```{r combine_dnase_chip_data, eval=FALSE}
# tf_name: TF name.
# cell_type: cell type.
# metadata: Filename for training metadata table.
# thresh_pValue: FIMO p-value threshold. 
# bin: type of MILLIPEDE binning scheme (Default: 'M5').
# chip_counts: ChIP read counts.
# dnase_dir: Directory for DNase or ATAC count matrices.
# dnase_idxstats_dir: Directory for DNase or ATAC reads idxstats files.
# outdir: Output directory.
# outname: Output filename prefix.
Rscript ../scripts/combine_dnase_chip_data.R \
        --tf_name tf_name \
        --cell_type cell_type \
        --metadata metadata \
        --thresh_pValue 1e-5 \
        --bin 'M5' \
        --chip_counts chip_counts \
        --dnase_dir dnase_dir \
        --dnase_idxstats_dir dnase_idxstats_dir \
        --outdir outdir \
        --outname outname
```

Add ChIP binary labels (from ChIP-seq peaks) to the combined training data
```{r add_chip_labels_to_combined_data, eval=FALSE}
# tf_name: TF name.
# cell_type: cell type.
# metadata: Filename for training metadata table.
# thresh_pValue: FIMO p-value threshold. 
# bin: type of MILLIPEDE binning scheme (Default: 'M5').
# combined_data_dir: Directory for combined data for each TF x cell type combo.
# chip_peak_dir: Directory for ChIP-seq peak bed narrowPeak files.
Rscript ../scripts/add_chip_peaks_to_combined_data.R \
        --tf_name tf_name \
        --cell_type cell_type \
        --metadata metadata \
        --thresh_pValue 1e-5 \
        --bin 'M5' \
        --combined_data_dir combined_data_dir \
        --chip_peak_dir chip_peak_dir
```

## Snakemake pipeline
We provided a [Snakemake pipeline](https://github.com/HarteminkLab/TOP/tree/main/snakemake) to automate the whole process and helpful if you have many TFs (motifs) in many cell types.

Run `Snakefile_training_ATAC` for ATAC-seq, or `Snakefile_training_DNase` for DNase-seq. For more details and instructions about running Snakemake pipelines, see [this page](https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html).
