<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Prepare input data • TOP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Prepare input data">
<meta property="og:description" content="TOP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.13.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_model.html">Train TOP quantitative occupancy model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_logistic_model.html">Train TOP logistic (binary binding) model</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HarteminkLab/TOP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Prepare input data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/HarteminkLab/TOP/blob/HEAD/vignettes/data_preparation.Rmd" class="external-link"><code>vignettes/data_preparation.Rmd</code></a></small>
      <div class="hidden name"><code>data_preparation.Rmd</code></div>

    </div>

    
    
<p>TOP predicts the quantitative TF occupancy from ChIP-seq around candidate TF binding sites using the site-centric approach.</p>
<p>We first identified candidate binding sites by motif scanning with a permissive threshold (using FIMO) (Grant et al., 2011).</p>
<p>Then, for each cell type, we considered (normalized) DNase and/or ATAC cleavage events occurring within 100 bp of the candidate binding site.</p>
<p>Similarly, we quantified TF occupancy in terms of ChIP-seq read counts within 100 bp of the candidate binding site, and this served as the target of our regression when training TOP.</p>
<p>We simplified the chromatin accessibility data into predictive features using five bins that aggregate the number of cleavage events occurring within the motif itself, as well as within two non-overlapping flanking regions upstream and downstream, using the same binning scheme used in the MILLIPEDE model (Luo and Hartemink, 2013).</p>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>The input data for TOP includes a data frame for each TF x cell type of interest:</p>
<p>The input data frame (see example below) should contains:</p>
<ul>
<li><p>Columns of the candidate binding sites: chr, start, end, site name, PWM score, strand, p-value from (FIMO) motif scanning.</p></li>
<li><p>Columns of DNase-seq or ATAC-seq bin counts: Five bins (MILLIPEDE M5 bins) around motif matches.</p></li>
<li><p>optional: one column of ChIP-seq measurement, if you want to train your own models. It could be quantitative TF occupancy (asinh transformed ChIP-seq read counts) or binary TF binding labels (obtained using ChIP-seq peaks).</p></li>
</ul>
<p>You can follow the procedure below or use your own data processing scripts/pipelines to prepare the input data.</p>
<p>You can also include additional steps to correct DNase-seq or ATAC-seq bias. Yet, our binning approach (with M5 bins) should make TOP robust to DNase-seq or ATAC-seq cleavage bias at base-pair resolution.</p>
<p>We recommend using <a href="https://snakemake.readthedocs.io/en/stable/" class="external-link">Snakemake</a> to automate the whole procedure, especially when you have data for many TFs in many cell types or conditions. We provided some <a href="https://github.com/HarteminkLab/TOP-paper-resources/tree/main/code/snakemake" class="external-link">example pipelines</a> that we used in our paper.</p>
<p>In order to prepare input data using <code>TOP</code> functions, please have the following R packages installed: <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html" class="external-link">GenomicRanges</a>, <a href="https://bioconductor.org/packages/release/bioc/html/Rsamtools.html" class="external-link">Rsamtools</a>, <a href="https://cran.r-project.org/web/packages/data.table/index.html" class="external-link">data.table</a>, <a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a>, as well as the following command line tools: <a href="https://bedtools.readthedocs.io/en/latest/" class="external-link">bedtools</a>, <a href="https://github.com/CRG-Barcelona/bwtool/wiki" class="external-link">bwtool</a>, and <a href="https://meme-suite.org/meme/doc/fimo.html" class="external-link">fimo</a> from <a href="https://meme-suite.org/meme/index.html" class="external-link">the MEME suite</a> as well as <a href="https://genome.ucsc.edu/goldenpath/help/bigWig.html" class="external-link">bedGraphToBigWig</a> and <a href="https://genome.ucsc.edu/goldenpath/help/bigWig.html" class="external-link">bigWigAverageOverBed</a> from <a href="http://hgdownload.soe.ucsc.edu/admin/exe/" class="external-link">UCSC binary utilities directory</a>.</p>
</div>
<div class="section level2">
<h2 id="example-data-preparation-procedure">Example data preparation procedure<a class="anchor" aria-label="anchor" href="#example-data-preparation-procedure"></a>
</h2>
<p>Here, we show an example procedure with several steps for preparing input data for CTCF in K562 cell type using the human genome assembly version hg38.</p>
<p>Load TOP R package</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HarteminkLab/TOP" class="external-link">TOP</a></span><span class="op">)</span></code></pre></div>
<p><strong>Step 1: Find TF motif matches using FIMO software</strong></p>
<p>To scan for TF motif matches, we use the <a href="https://meme-suite.org/meme/doc/fimo.html" class="external-link">FIMO</a> software from the MEME suite.</p>
<p>Download hg38 reference genome FASTA file and save it as <code>hg38.fa</code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/analysisSet/hg38.analysisSet.fa.gz</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> <span class="at">-c</span> hg38.analysisSet.fa.gz <span class="op">&gt;</span> hg38.fa</span></code></pre></div>
<p>Generate the <code>chrom.sizes</code> file which will be needed later.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/index_fa.html">index_fa</a></span><span class="op">(</span><span class="st">'hg38.fa'</span>, chromsize_file<span class="op">=</span><span class="st">'hg38.chrom.sizes'</span><span class="op">)</span></code></pre></div>
<p>Download the CTCF motif file <code>MA0139.1.meme</code> (in MEME format) from <a href="https://jaspar.genereg.net/matrix/MA0139.1/" class="external-link">JASPAR</a>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://jaspar.genereg.net/api/v1/matrix/MA0139.1.meme</span></code></pre></div>
<p>We use FIMO with the threshold <code>p-value &lt; 1e-5</code> in most of our analyses. You can set <code>thresh = "1e-4"</code> (FIMO’s default threshold), if you need more motif matches.</p>
<p>You can also set different background model in <code>bfile</code>, for example, <code>bfile='--uniform--'</code> would use uniform nucleotide frequencies.</p>
<p>Save results in <code>MA0139.1_1e-5.fimo.txt</code> and <code>MA0062.1_1e-5.fimo.txt</code>, which will be used in the next step.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/fimo_motif_matches.html">fimo_motif_matches</a></span><span class="op">(</span>motif_file<span class="op">=</span><span class="st">'MA0139.1.meme'</span>, 
                   sequence_file<span class="op">=</span><span class="st">'hg38.fa'</span>, 
                   thresh_pValue<span class="op">=</span><span class="fl">1e-5</span>, 
                   outname<span class="op">=</span><span class="st">'MA0139.1_1e-5.fimo.txt'</span><span class="op">)</span></code></pre></div>
<p><strong>Step 2: Get candidate TF binding sites</strong></p>
<p>We take motif matches obtained from FIMO as candidate binding sites, and add 100 bp flanking regions on both sides of the motifs, then filter candidate sites by FIMO p-value, and filter the candidate sites falling in ENCODE blacklist regions.</p>
<p>Download ENCODE blacklist from <a href="https://www.encodeproject.org/annotations/ENCSR636HFF/" class="external-link">ENCODE portal</a> and save as <code>blacklist.hg38.bed.gz</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ENCFF356LFX.bed.gz blacklist.hg38.bed.gz</span></code></pre></div>
<p>Obtain the candidate sites</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># fimo_file: FIMO result file.</span>
<span class="co"># thresh_pValue: FIMO p-value threshold.</span>
<span class="co"># blacklist_file: file with ENOCDE blacklist regions.</span>
<span class="va">sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_candidate_sites.html">process_candidate_sites</a></span><span class="op">(</span>fimo_file<span class="op">=</span><span class="st">'MA0139.1_1e-5.fimo.txt'</span>, 
                                 thresh_pValue<span class="op">=</span><span class="fl">1e-5</span>, 
                                 blacklist_file<span class="op">=</span><span class="st">'blacklist.hg38.bed.gz'</span><span class="op">)</span></code></pre></div>
<p><strong>Step 3: Count DNase- or ATAC-seq genome-wide cleavage</strong></p>
<p>We use ATAC-seq reads from K562 cell line (ENCODE ID: <code>ENCSR868FGK</code>) for example.</p>
<p>There are three replicate samples in this study. Here we only use one replicate sample (<code>ENCFF534DCE.bam</code>, and we renamed it as <code>K562.ATAC.bam</code>) as an example.</p>
<p>We first sort and index the BAM file, and obtain the total number of mapped reads from the idxstats file, which will be used later when normalizing read counts by library sizes.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the BAM file from ENCODE</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://www.encodeproject.org/files/ENCFF534DCE/@@download/ENCFF534DCE.bam</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename the bam file</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ENCFF534DCE.bam K562.ATAC.bam</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This BAM file has already been sorted, so we skip the sorting step. </span>
<span class="fu"><a href="../reference/sort_index_idxstats_bam.html">sort_index_idxstats_bam</a></span><span class="op">(</span><span class="st">'K562.ATAC.bam'</span>, sort<span class="op">=</span><span class="cn">FALSE</span>, index<span class="op">=</span><span class="cn">TRUE</span>, idxstats<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Next, we count the DNase or ATAC cuts along the genome, and save the genome counts in BigWig files. This step often takes a while especially for large BAM files, but it only needs to be done once.</p>
<p>The resulting BigWig files will be used later for extracting the count matrices around the candidate sites for different motifs.</p>
<p>We need the command line tools: <a href="https://bedtools.readthedocs.io/en/latest/" class="external-link"><code>bedtools</code></a> and <code>bedGraphToBigWig</code> from <a href="http://hgdownload.soe.ucsc.edu/admin/exe/" class="external-link">UCSC binary utilities</a> for this step.</p>
<p>For ATAC-seq, when , reads were shifted so as to address offsets and align the signal across strands.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># bam_file: sorted BAM file.</span>
<span class="co"># chrom_size_file: file of genome sizes by chromosomes.</span>
<span class="co"># shift_ATAC: If TRUE, shift ATAC-seq reads.</span>
<span class="co"># outdir: directory for saving the BigWig files of genome counts, same as outdir in get_sites_counts().</span>
<span class="co"># outname: prefix for the BigWig files, same as genomecount_name in get_sites_counts().</span>
<span class="fu"><a href="../reference/count_genome_cuts.html">count_genome_cuts</a></span><span class="op">(</span>bam_file<span class="op">=</span><span class="st">'K562.ATAC.bam'</span>, 
                  chrom_size_file<span class="op">=</span><span class="st">'hg38.chrom.sizes'</span>, 
                  shift_ATAC<span class="op">=</span><span class="cn">TRUE</span>,
                  outdir<span class="op">=</span><span class="st">'processed_data'</span>,
                  outname<span class="op">=</span><span class="st">'K562.ATAC'</span><span class="op">)</span></code></pre></div>
<p><strong>Step 4: Get DNase- or ATAC-seq count matrices around candidate sites, then normalize, bin and transform the counts</strong></p>
<p>Get DNase- or ATAC-seq read counts matrix around candidate sites:</p>
<p>We utilize <a href="https://pubmed.ncbi.nlm.nih.gov/24489365/" class="external-link">bwtool</a> to extract read counts from the BigWig files generated earlier.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># sites: data frame containing candidate sites.</span>
<span class="co"># genomecount_dir: directory for genome counts, same as outdir in count_genome_cuts().</span>
<span class="co"># genomecount_name: file prefix for genome counts, same as outname in count_genome_cuts().</span>
<span class="va">sites_counts.mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_sites_counts.html">get_sites_counts</a></span><span class="op">(</span><span class="va">sites</span>,
                                     genomecount_dir<span class="op">=</span><span class="st">'processed_data'</span>,
                                     genomecount_name<span class="op">=</span><span class="st">'K562.ATAC'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">sites_counts.mat</span>, <span class="st">"processed_data/CTCF.K562.ATAC.counts.mat.rds"</span><span class="op">)</span></code></pre></div>
<p>Normalize, bin and transform counts:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># count_matrix: DNase (or ATAC) read counts matrix.</span>
<span class="co"># idxstats_file: the idxstats file generated by samtools.</span>
<span class="co"># ref_size: scale to reference library size.</span>
<span class="co"># (default: 50 million for ATAC-seq, 100 million for DNase-seq)</span>
<span class="co"># bin_method: MILLIPEDE binning scheme (default: 'M5').</span>
<span class="co"># transform: type of transformation for DNase (or ATAC) counts (default: 'asinh').</span>
<span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize_bin_transform_counts.html">normalize_bin_transform_counts</a></span><span class="op">(</span><span class="va">sites_counts.mat</span>, 
                                       idxstats_file<span class="op">=</span><span class="st">'K562.ATAC.bam.idxstats.txt'</span>, 
                                       ref_size<span class="op">=</span><span class="fl">5e7</span>,
                                       bin_method<span class="op">=</span><span class="st">'M5'</span>,
                                       transform<span class="op">=</span><span class="st">'asinh'</span><span class="op">)</span></code></pre></div>
<p>Make a data frame for candidate sites combining motif match information and transformed ATAC (or DNase) counts in five MILLIPEDE bins:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">sites</span>, <span class="va">bins</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">combined_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'name'</span>,<span class="st">'pwm.score'</span>,<span class="st">'strand'</span>,<span class="st">'p.value'</span>, 
                                <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'bin'</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">combined_data</span>, <span class="st">'processed_data/CTCF_MA0139.1_1e-5.K562.ATAC.M5.combined.data.rds'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined_data</span>, <span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;    chr start   end  name pwm.score strand  p.value      bin1      bin2 bin3</span>
<span class="co">#&gt; 1 chr1 11122 11341 site1   22.4839      - 9.17e-09 0.2232134 0.0000000    0</span>
<span class="co">#&gt; 2 chr1 11180 11399 site2   20.7581      - 5.22e-08 0.0000000 0.2232134    0</span>
<span class="co">#&gt; 3 chr1 24681 24900 site3   16.5645      - 1.31e-06 0.0000000 0.0000000    0</span>
<span class="co">#&gt;   bin4 bin5</span>
<span class="co">#&gt; 1    0    0</span>
<span class="co">#&gt; 2    0    0</span>
<span class="co">#&gt; 3    0    0</span></code></pre></div>
<p>We can apply TOP models to data frames like this to make predictions.</p>
<p><strong>Prepare ChIP-seq data (optional if you want to train your own model)</strong></p>
<p>If you want to train your own model, you would also need to prepare ChIP data and add those to your input data.</p>
<p>Download CTCF K562 ChIP-seq BAM files (ENCODE ID: <code>ENCSR000EGM</code>, two replicates: <code>ENCFF172KOJ</code> and <code>ENCFF265ZSP</code>).</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the ChIP-seq BAM files</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://www.encodeproject.org/files/ENCFF172KOJ/@@download/ENCFF172KOJ.bam</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://www.encodeproject.org/files/ENCFF265ZSP/@@download/ENCFF265ZSP.bam</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename the BAM files</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ENCFF172KOJ.bam CTCF.K562.ChIPseq.rep1.bam</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ENCFF265ZSP.bam CTCF.K562.ChIPseq.rep2.bam</span></code></pre></div>
<p>Sort and index the BAM files and obtain the number of mapped reads.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># The BAM files have already been sorted, so we skip the sorting step. </span>
<span class="fu">bam_sort_index_stats</span><span class="op">(</span><span class="st">'CTCF.K562.ChIPseq.rep1.bam'</span>, sort<span class="op">=</span><span class="cn">FALSE</span>, index<span class="op">=</span><span class="cn">TRUE</span>, idxstats<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">bam_sort_index_stats</span><span class="op">(</span><span class="st">'CTCF.K562.ChIPseq.rep2.bam'</span>, sort<span class="op">=</span><span class="cn">FALSE</span>, index<span class="op">=</span><span class="cn">TRUE</span>, idxstats<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Count ChIP-seq reads around candidate sites (merge ChIP-seq replicates), and normalize (scale) to the same reference library size (20 million).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># sites: data frame containing the candidate sites.</span>
<span class="co"># chip_bam_files: ChIP-seq bam files (may include multiple replicates).</span>
<span class="co"># chrom_size_file: file of genome sizes by chromosomes.</span>
<span class="co"># ref_size: normalize to ChIP-Seq reference library size (default: 20 million).</span>
<span class="va">sites_chip</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_normalize_chip.html">count_normalize_chip</a></span><span class="op">(</span><span class="va">sites</span>,
                                   chip_bam_files<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'CTCF.K562.ChIPseq.rep1.bam'</span>,
                                                    <span class="st">'CTCF.K562.ChIPseq.rep2.bam'</span><span class="op">)</span>,
                                   chrom_size_file<span class="op">=</span><span class="st">'hg38.chrom.sizes'</span>, 
                                   ref_size<span class="op">=</span><span class="fl">2e7</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sites_chip</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p>Combine motif match information, ATAC (or DNase) bins, and ChIP-seq counts into a data frame.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">sites</span>, <span class="va">bins</span>, chip <span class="op">=</span> <span class="va">sites_chip</span><span class="op">$</span><span class="va">chip</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">combined_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'name'</span>,<span class="st">'pwm.score'</span>,<span class="st">'strand'</span>,<span class="st">'p.value'</span>, 
                                <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'bin'</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span><span class="op">)</span>, <span class="st">'chip'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">combined_data</span>, <span class="st">'processed_data/CTCF_MA0139.1_1e-5.K562.ATAC.M5.ChIP.combined.data.rds'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined_data</span>, <span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;    chr start   end  name pwm.score strand  p.value      bin1      bin2 bin3</span>
<span class="co">#&gt; 1 chr1 11122 11341 site1   22.4839      - 9.17e-09 0.2232134 0.0000000    0</span>
<span class="co">#&gt; 2 chr1 11180 11399 site2   20.7581      - 5.22e-08 0.0000000 0.2232134    0</span>
<span class="co">#&gt; 3 chr1 24681 24900 site3   16.5645      - 1.31e-06 0.0000000 0.0000000    0</span>
<span class="co">#&gt;   bin4 bin5 chip</span>
<span class="co">#&gt; 1    0    0    0</span>
<span class="co">#&gt; 2    0    0    0</span>
<span class="co">#&gt; 3    0    0    0</span></code></pre></div>
<p>Replace ChIP-seq counts with binary ChIP labels (from ChIP-seq peaks) if you want to train TOP logistic version to predict TF binding probability.</p>
<p>Download CTCF K562 ChIP-seq peaks</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download ChIP-seq peaks</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://www.encodeproject.org/files/ENCFF660GHM/@@download/ENCFF660GHM.bed.gz</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename the BAM files</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ENCFF660GHM.bed.gz CTCF.K562.ChIPseq.peaks.bed.gz</span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># We can directly supply the ChIP-seq peaks file name in 'chip_peak_file',</span>
<span class="co"># alternatively, we can set 'chip_peak_sampleID'. Then, it will download the </span>
<span class="co"># ChIP-seq peaks file from ENCODE website.</span>
<span class="va">sites_chip_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_chip_peak_labels_to_sites.html">add_chip_peak_labels_to_sites</a></span><span class="op">(</span><span class="va">sites</span>, 
                                                   chip_peak_file<span class="op">=</span><span class="st">'CTCF.K562.ChIPseq.peaks.bed.gz'</span><span class="op">)</span>

<span class="va">combined_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">sites</span>, <span class="va">bins</span>, chip_label <span class="op">=</span> <span class="va">sites_chip_labels</span><span class="op">$</span><span class="va">chip_label</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">combined_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'name'</span>,<span class="st">'pwm.score'</span>,<span class="st">'strand'</span>,<span class="st">'p.value'</span>, 
                             <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'bin'</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span><span class="op">)</span>, <span class="st">'chip_label'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">combined_data</span>, <span class="st">'processed_data/CTCF_MA0139.1_1e-5.K562.ATAC.M5.ChIPlabels.combined.data.rds'</span><span class="op">)</span></code></pre></div>
<p>We can plot the footprint profiles around the motif matches</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bound_sites_counts.mat</span> <span class="op">&lt;-</span> <span class="va">sites_counts.mat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">combined_data</span><span class="op">$</span><span class="va">chip_label</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span>, <span class="op">]</span>
<span class="fu"><a href="../reference/plot_profile.html">plot_profile</a></span><span class="op">(</span><span class="va">bound_sites_counts.mat</span>, title<span class="op">=</span><span class="st">'ATAC-seq profiles around CTCF binding sites'</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
