<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Prepare input data • TOP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Prepare input data">
<meta property="og:description" content="TOP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_model.html">Train TOP model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_logistic_model.html">Train TOP logistic model</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinlkx/TOP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Prepare input data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kevinlkx/TOP/blob/HEAD/vignettes/data_preparation.Rmd" class="external-link"><code>vignettes/data_preparation.Rmd</code></a></small>
      <div class="hidden name"><code>data_preparation.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>TOP requires a data frame for each TF and cell type of interest.</p>
<p>The format of the input data frame (see example data below):</p>
<ul>
<li><p>Columns of the candidate binding sites: chr, start, end, site name, PWM score, strand, p-value from (FIMO) motif match scanning.</p></li>
<li><p>Columns of DNase-seq or ATAC-seq bins: MILLIPEDE bins around motif matches (default: M5 bins).</p></li>
<li><p>optional: one column of ChIP-seq measurement, if you want to train the models. It could be quantitative TF occupancy (asinh transformed ChIP-seq read counts) or binary TF binding labels (from ChIP-seq peaks).</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="example-procedure">Example procedure<a class="anchor" aria-label="anchor" href="#example-procedure"></a>
</h2>
<p>Here are example steps of preparing input data for TOP to predict CTCF occupancy in K562 cell type:</p>
<p><strong>Step 1: Find TF motif matches using FIMO software</strong></p>
<p>To scan for TF motif matches, we use the <a href="http://meme-suite.org/doc/fimo.html?man_type=web" class="external-link">FIMO</a> software from the MEME suite.</p>
<p>We use the command line version of FIMO. By default, we used with the threshold <code>p-value &lt; 1e-5</code> and uniform background nucleotide frequency. You can use FIMO’s default threshold <code>p-value &lt; 1e-4</code> (which will result in more motif matches), or use other background nucleotide frequency.</p>
<p>FIMO command line: <code>fimo [options] &lt;motif file&gt; &lt;sequence file&gt;</code>.</p>
<ul>
<li>motif file: The name of a file containing MEME formatted motifs.</li>
<li>sequence file: The name of a file containing a collection of sequences in FASTA format.</li>
</ul>
<p>Download the CTCF motif file <code>MA0139.1.meme</code> (in MEME format) from <a href="https://jaspar.genereg.net/matrix/MA0139.1/" class="external-link">JASPAR</a>.</p>
<p>Download hg38 fasta file and save as <code>hg38.fa</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/analysisSet/hg38.analysisSet.fa.gz</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> hg38.analysisSet.fa.gz</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> hg38.analysisSet.fa hg38.fa</span></code></pre></div>
<p>Generate the chrom.sizes file which will be needed later.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx hg38.fa</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1,2 hg38.fa.fai <span class="op">&gt;</span> hg38.chrom.sizes</span></code></pre></div>
<p>Save FIMO output in text format (<code>MA0139.1_1e-5.fimo.txt</code>), which will be used in the next step.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find motif matches for CTCF</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fimo</span> <span class="at">--text</span> <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>--skip-matched-sequence <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>--verbosity 2 <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>--bgfile <span class="at">--uniform--</span> <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>--thresh 1e-5 <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>--max-stored-scores 1000000 <span class="dt">\</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>MA0139.1.meme hg38.fa <span class="dt">\</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> MA0139.1_1e-5.fimo.txt</span></code></pre></div>
<p><strong>Step 2: Get candidate TF binding sites</strong></p>
<p>We take motif matches obtained from FIMO (<code>MA0139.1_1e-5.fimo.txt</code>) as candidate binding sites, and add 100 bp flanking regions on both sides of the motifs, then filter candidate sites by FIMO p-value, and filter the candidate sites falling in ENCODE blacklist regions. We save the candidate sites in a text file (<code>MA0139.1_1e-5.candidate_sites.txt</code>)</p>
<p>Download ENCODE blacklist from <a href="https://www.encodeproject.org/annotations/ENCSR636HFF/" class="external-link">ENCODE portal</a></p>
<p>Load TOP R package</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinlkx/TOP" class="external-link">TOP</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sites.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_candidate_sites.html">process_candidate_sites</a></span><span class="op">(</span>fimo_file<span class="op">=</span><span class="st">'MA0139.1_1e-5.fimo.txt'</span>, 
                                    flank<span class="op">=</span><span class="fl">100</span>, 
                                    thresh_pValue<span class="op">=</span><span class="fl">1e-5</span>, 
                                    blacklist_file<span class="op">=</span><span class="st">'blacklist.hg38.bed.gz'</span><span class="op">)</span></code></pre></div>
<p><strong>Step 3: Count DNase- or ATAC-seq genome-wide cleavage</strong></p>
<p>We use ATAC-seq reads in K562 cell line from ENCODE (ID: <code>ENCSR868FGK</code>) for example.</p>
<p>There are three replicate samples in this study. Here we only use one replicate sample (<code>ENCFF534DCE.bam</code>) as an example, and rename the file as <code>K562.ATAC.bam</code>.</p>
<p>We first sort and index the BAM file using <code>samtools</code>, and then print the stats of the reads, which will be used later when normalizing read counts by library sizes.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rename the bam file</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ENCFF534DCE.bam K562.ATAC.bam</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># This bam file is already sorted, so we skip the sorting step. </span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index K562.ATAC.bam</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> idxstats K562.ATAC.bam <span class="op">&gt;</span> K562.ATAC.bam.idxstats.txt</span></code></pre></div>
<p>Next, we count the cleavages along the genome, and save in Bigwig format. This step takes more time but only needs to be done once, and allows us to efficiently extract the read counts around candidate sites for many different motifs.</p>
<p>We need <code>bedtools</code> and <code>bedGraphToBigWig</code> from UCSC for this step.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># bam_file: input BAM filename.</span>
<span class="co"># chrom_size_file: chrom sizes file.</span>
<span class="co"># outdir: Save bigWig files of genome counts in this directory.</span>
<span class="fu"><a href="../reference/count_genome_cuts.html">count_genome_cuts</a></span><span class="op">(</span>bam_file <span class="op">=</span> <span class="st">'K562.ATAC.bam'</span>, 
                  chrom_size_file <span class="op">=</span> <span class="st">'hg38.chrom.sizes'</span>, 
                  outdir <span class="op">=</span> <span class="st">"processed_data/"</span><span class="op">)</span></code></pre></div>
<p><strong>Step 4: Get DNase- or ATAC-seq count matrices for each motif, then normalize, bin and transform the counts</strong></p>
<p>Get count matrices around candidate sites. We need <a href="https://pubmed.ncbi.nlm.nih.gov/24489365/" class="external-link">bwtool</a> for this step.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># sites_file: Candidate sites filename.</span>
<span class="co"># genomecount_dir: Genome counts directory.</span>
<span class="co"># genomecount_name: Genome counts prefix.</span>
<span class="va">sites_counts.mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_sites_counts.html">get_sites_counts</a></span><span class="op">(</span><span class="va">sites.df</span>, 
                                     genomecount_dir<span class="op">=</span><span class="st">'./processed_data/'</span>,
                                     genomecount_name<span class="op">=</span><span class="st">'K562.ATAC'</span><span class="op">)</span></code></pre></div>
<p>Normalize, bin and transform counts.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># count_matrix: ATAC (or DNase) count matrix</span>
<span class="co"># idxstats_file: The idxstats file generated by samtools</span>
<span class="co"># ref.size: Scale to reference library size </span>
<span class="co"># (default: 50 million for ATAC-seq, 100 million for DNase-seq)</span>
<span class="co"># bin.method: MILLIPEDE binning scheme (default: 'M5').</span>
<span class="co"># transform: asinh transform of the counts.</span>
<span class="va">bins.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize_bin_transform_counts.html">normalize_bin_transform_counts</a></span><span class="op">(</span><span class="va">sites_counts.mat</span>, 
                                          idxstats_file<span class="op">=</span><span class="st">'K562.ATAC.bam.idxstats.txt'</span>, 
                                          ref.size<span class="op">=</span><span class="fl">5e7</span>,
                                          bin.method<span class="op">=</span><span class="st">'M5'</span>,
                                          transform<span class="op">=</span><span class="st">'asinh'</span><span class="op">)</span></code></pre></div>
<p>Make a data frame with PWM scores, and ATAC (or DNase) bins. We can apply TOP model on this data frame to make predictions.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_data.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">sites.df</span>, <span class="va">bins.df</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">combined_data.df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'name'</span>,<span class="st">'pwm.score'</span>,<span class="st">'strand'</span>,<span class="st">'p.value'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'bin'</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bins.df</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">combined_data.df</span>, <span class="st">'processed_data/CTCF_MA0139.1_1e-5.K562.ATAC.M5.combined.data.rds'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_data.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'../inst/example/processed_data/CTCF_MA0139.1_1e-5.K562.ATAC.M5.combined.data.rds'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined_data.df</span>, <span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;    chr start   end  name pwm.score strand  p.value      bin1      bin2 bin3</span>
<span class="co">#&gt; 1 chr1 11122 11341 site1   22.4839      - 9.17e-09 0.2232134 0.0000000    0</span>
<span class="co">#&gt; 2 chr1 11180 11399 site2   20.7581      - 5.22e-08 0.0000000 0.2232134    0</span>
<span class="co">#&gt; 3 chr1 24681 24900 site3   16.5645      - 1.31e-06 0.0000000 0.0000000    0</span>
<span class="co">#&gt;   bin4 bin5</span>
<span class="co">#&gt; 1    0    0</span>
<span class="co">#&gt; 2    0    0</span>
<span class="co">#&gt; 3    0    0</span></code></pre></div>
<p><strong>Step 5: Prepare ChIP-seq data (optional if you want to train your own model)</strong></p>
<p>Download CTCF K562 ChIP-seq bam files (ENCODE ID: <code>ENCSR000EGM</code>, two replicates: <code>ENCFF172KOJ</code> and <code>ENCFF265ZSP</code>). We again sort and index the BAM files and print the stats of the reads using <code>samtools</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rename the bam files</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ENCFF172KOJ.bam CTCF.K562.ChIPseq.rep1.bam</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ENCFF265ZSP.bam CTCF.K562.ChIPseq.rep2.bam</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># The bam files are already sorted, so we skip the sorting step. </span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index CTCF.K562.ChIPseq.rep1.bam</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> idxstats CTCF.K562.ChIPseq.rep1.bam <span class="op">&gt;</span> CTCF.K562.ChIPseq.rep1.bam.idxstats.txt</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index CTCF.K562.ChIPseq.rep2.bam</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> idxstats CTCF.K562.ChIPseq.rep2.bam <span class="op">&gt;</span> CTCF.K562.ChIPseq.rep2.bam.idxstats.txt</span></code></pre></div>
<p>Count ChIP-seq reads around candidate sites (merge ChIP-seq replicates), and normalize (scale) to the same reference library size (10 million).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># ref.size: ChIP-Seq reference library size (default: 10 million)</span>
<span class="va">sites_chip.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_normalize_chip.html">count_normalize_chip</a></span><span class="op">(</span><span class="va">sites.df</span>,
                                      chip_bam_files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'CTCF.K562.ChIPseq.rep1.bam'</span>, <span class="st">'CTCF.K562.ChIPseq.rep2.bam'</span><span class="op">)</span>,
                                      chrom_size_file <span class="op">=</span> <span class="st">'hg38.chrom.sizes'</span>, 
                                      ref.size <span class="op">=</span> <span class="fl">1e7</span><span class="op">)</span></code></pre></div>
<p>Combine PWM scores, ATAC (or DNase) bins, and ChIP-seq counts into a data frame.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_data.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">sites.df</span>, <span class="va">bins.df</span>, <span class="va">sites_chip.df</span><span class="op">$</span><span class="va">chip</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">combined_data.df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'name'</span>,<span class="st">'pwm.score'</span>,<span class="st">'strand'</span>,<span class="st">'p.value'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'bin'</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bins.df</span><span class="op">)</span><span class="op">)</span>, <span class="st">'chip'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">combined_data.df</span>, <span class="st">'processed_data/CTCF_MA0139.1_1e-5.K562.ATAC.M5.ChIP.combined.data.rds'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_data.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'../inst/example/processed_data/CTCF_MA0139.1_1e-5.K562.ATAC.M5.ChIP.combined.data.rds'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined_data.df</span>, <span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;    chr start   end  name pwm.score strand  p.value      bin1      bin2 bin3</span>
<span class="co">#&gt; 1 chr1 11122 11341 site1   22.4839      - 9.17e-09 0.2232134 0.0000000    0</span>
<span class="co">#&gt; 2 chr1 11180 11399 site2   20.7581      - 5.22e-08 0.0000000 0.2232134    0</span>
<span class="co">#&gt; 3 chr1 24681 24900 site3   16.5645      - 1.31e-06 0.0000000 0.0000000    0</span>
<span class="co">#&gt;   bin4 bin5 chip</span>
<span class="co">#&gt; 1    0    0    0</span>
<span class="co">#&gt; 2    0    0    0</span>
<span class="co">#&gt; 3    0    0    0</span></code></pre></div>
<p>Replace ChIP-seq counts with binary ChIP labels (from ChIP-seq peaks) if you want to train TOP logistic version to predict TF binding probability.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sites_chip_labels.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_chip_peak_labels_to_sites.html">add_chip_peak_labels_to_sites</a></span><span class="op">(</span><span class="va">sites.df</span>, 
                                                      chip_peak_sampleID <span class="op">=</span> <span class="st">'ENCFF660GHM'</span>,
                                                      chip_peak_dir <span class="op">=</span> <span class="st">'./processed_data/'</span><span class="op">)</span>
<span class="va">combined_data.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">sites.df</span>, <span class="va">bins.df</span>, <span class="va">sites_chip_labels.df</span><span class="op">$</span><span class="va">chip_label</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">combined_data.df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'name'</span>,<span class="st">'pwm.score'</span>,<span class="st">'strand'</span>,<span class="st">'p.value'</span>, 
                                <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'bin'</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bins.df</span><span class="op">)</span><span class="op">)</span>, <span class="st">'chip_label'</span><span class="op">)</span></code></pre></div>
<p>You can also replace ChIP-seq counts with signal values (fold-over control at each position from ENCODE).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sites_chip_signals.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_chip_signals_to_sites.html">add_chip_signals_to_sites</a></span><span class="op">(</span><span class="va">sites.df</span>, 
                                                   chip_signal_sampleID <span class="op">=</span> <span class="st">'ENCFF682MFJ'</span>,
                                                   chip_signal_dir <span class="op">=</span> <span class="st">'./processed_data/'</span><span class="op">)</span>

<span class="va">combined_data.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">sites.df</span>, <span class="va">bins.df</span>, <span class="va">sites_chip_signals.df</span><span class="op">$</span><span class="va">chip_signal</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">combined_data.df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'name'</span>,<span class="st">'pwm.score'</span>,<span class="st">'strand'</span>,<span class="st">'p.value'</span>, 
                                <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'bin'</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bins.df</span><span class="op">)</span><span class="op">)</span>, <span class="st">'chip_signal'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="snakemake-pipeline">Snakemake pipeline<a class="anchor" aria-label="anchor" href="#snakemake-pipeline"></a>
</h2>
<p>We recommend writing a Snakemake pipeline to automate the whole procedure. It is especially helpful if you have many TFs and many cell types.</p>
<p>For more details and instructions about Snakemake pipelines, see <a href="https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html" class="external-link">Snakemake tutorial</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
