<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An example for input data preparation • TOP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An example for input data preparation">
<meta property="og:description" content="TOP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/data_preparation_example.html">An example for input data preparation</a>
    </li>
    <li>
      <a href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_model_JAGS.html">Train TOP model</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinlkx/TOP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An example for input data preparation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kevinlkx/TOP/blob/HEAD/vignettes/data_preparation_example.Rmd" class="external-link"><code>vignettes/data_preparation_example.Rmd</code></a></small>
      <div class="hidden name"><code>data_preparation_example.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="top-input-data">TOP input data<a class="anchor" aria-label="anchor" href="#top-input-data"></a>
</h2>
<p>TOP requires a data frame as input data for each TF in each cell type.</p>
<p>The format of the input data frame:</p>
<ul>
<li>The first six columns: chr, start, end, site name, strand, motif PWM score.</li>
<li>The next five columns (if using M5 bins): five MILLIPEDE bins around motif matches.</li>
<li>optional: one ChIP column as the response variable. Could be quantitative TF occupancy (transformed ChIP-seq read counts) or binary TF binding labels (from ChIP-seq peaks).</li>
</ul>
<p>Here we give an example for the process of input data preparation for predicting CTCF and NRF1 occupancy in K562 cell type.</p>
</div>
<div class="section level2">
<h2 id="major-steps">Major steps<a class="anchor" aria-label="anchor" href="#major-steps"></a>
</h2>
<p><strong>Step 1: Find TF motif matches using FIMO software</strong></p>
<p>To scan for TF motif matches, we use the <a href="http://meme-suite.org/doc/fimo.html?man_type=web" class="external-link">FIMO</a> software from the MEME suite.</p>
<p>We use the command line version of FIMO. By default, we used with the threshold <span class="math inline">\(p &lt; 1e-5\)</span> and uniform background when training the model. You can choose your own background or use FIMO’s default threshold <span class="math inline">\(p &lt; 1e-4\)</span> (which will result in more motif matches).</p>
<p>Here we show an example for finding motif matches of NRF1.</p>
<p>FIMO command line: <code>fimo [options] &lt;motif file&gt; &lt;sequence file&gt;</code>. <motif file>: The name of a file containing MEME formatted motifs. <sequence file>: The name of a file containing a collection of sequences in FASTA format.</sequence></motif></p>
<p>Download the NRF1 motif file <code>MA0506.1.meme</code> (in MEME format) from <a href="https://jaspar.genereg.net/matrix/MA0506.1/" class="external-link">JASPAR</a>.</p>
<p>Download hg38 fasta file and save as <code>hg38.fa</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> hg38.fa.gz</span></code></pre></div>
<p>Save FIMO output in text format (MA0506.1_1e-5.fimo.txt), which will be used in the next step.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="va">topdir</span><span class="op">=</span>/project2/xinhe/kevinluo/TOP</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">${topdir}</span>/example</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Find motif matches for CTCF</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">fimo</span> <span class="at">--text</span> <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">--skip-matched-sequence</span> <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">--verbosity</span> 2 <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">--bgfile</span> <span class="at">--uniform--</span> <span class="dt">\</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">--thresh</span> 1e-5 <span class="dt">\</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">--max-stored-scores</span> 1000000 <span class="dt">\</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>     MA0139.1.meme hg38.fa <span class="dt">\</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>     <span class="op">&gt;</span> MA0139.1_1e-5.fimo.txt</span></code></pre></div>
<p><strong>Step 2: Get candidate TF binding sites</strong></p>
<p>We take motif matches obtained from FIMO (MA0506.1_1e-5.fimo.txt) as candidate binding sites, and add 100 bp flanking regions on both sides of the motifs, then filter candidate sites by FIMO p-value, and filter the candidate sites falling in ENCODE blacklist regions. We save the candidate sites in a text file (MA0506.1_1e-5.candidate_sites.txt)</p>
<p>Download ENCODE blacklist from <a href="https://www.encodeproject.org/annotations/ENCSR636HFF/" class="external-link">ENCODE portal</a></p>
<p><a href="https://github.com/kevinlkx/TOP/tree/main/inst/scripts/get_candidate_sites.R" class="external-link">Example script</a></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#module load R/4.0.4</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinlkx/TOP" class="external-link">TOP</a></span><span class="op">)</span>

<span class="va">blacklist_file</span> <span class="op">&lt;-</span> <span class="st">'../inst/extdata/blacklist.hg38.bed.gz'</span>

<span class="va">CTCF_sites.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_candidate_sites.html">process_candidate_sites</a></span><span class="op">(</span>fimo_file<span class="op">=</span><span class="st">'../inst/extdata/MA0139.1_1e-5.fimo.txt'</span>, 
                                         flank<span class="op">=</span><span class="fl">100</span>, 
                                         thresh_pValue<span class="op">=</span><span class="fl">1e-5</span>, 
                                         blacklist_file<span class="op">=</span><span class="va">blacklist_file</span>,
                                         out_file <span class="op">=</span> <span class="st">'../inst/extdata/MA0139.1_1e-5.candidate_sites.txt'</span><span class="op">)</span>
<span class="co">#&gt; Load FIMO result... </span>
<span class="co">#&gt; Flank motif matches by 100 bp... </span>
<span class="co">#&gt; Select candidate sites with FIMO p-value &lt; 1e-05 </span>
<span class="co">#&gt; Select candidate sites with PWM score &gt; 0 </span>
<span class="co">#&gt; Filter out 6633 sites in blacklist regions. </span>
<span class="co">#&gt; Save candidate sites at ../inst/extdata/MA0139.1_1e-5.candidate_sites.txt</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CTCF_sites.df</span><span class="op">)</span>
<span class="co">#&gt;   #chr  start   stop  name   score strand  p.value</span>
<span class="co">#&gt; 1 chr1  11122  11341 site1 22.4839      - 9.17e-09</span>
<span class="co">#&gt; 2 chr1  11180  11399 site2 20.7581      - 5.22e-08</span>
<span class="co">#&gt; 3 chr1  24681  24900 site3 16.5645      - 1.31e-06</span>
<span class="co">#&gt; 4 chr1  91319  91538 site4 15.1774      + 3.11e-06</span>
<span class="co">#&gt; 5 chr1 104884 105103 site5 15.6613      - 2.32e-06</span>
<span class="co">#&gt; 6 chr1 131227 131446 site6 13.1452      + 9.84e-06</span></code></pre></div>
<p><strong>Step 3: Count DNase- or ATAC-seq genome-wide cleavage</strong></p>
<p>We use ATAC-seq reads in K562 cell line from ENCODE (ID: ENCSR868FGK) for example.</p>
<p>Download ATAC-seq reads. There are three replicates in this study. We found modeling replicates do not improve our model, so we typically merge replicates.</p>
<p>For simplicity, here we only use one of the replicates (ENCFF534DCE.bam) as an example, and rename the file <code>ENCFF534DCE.bam</code> as <code>ATAC.K562.bam</code>.</p>
<p>We first sort and index the BAM file with DNase-seq or ATAC-seq reads using <code>samtools</code>, and then print the stats of the reads, which will be used later when normalizing read counts by library sizes.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This bam file is already sorted, so we skip the sorting step. </span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># samtools sort ATAC.K562.bam -o ATAC.K562.sorted.bam</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index ATAC.K562.bam</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> idxstats ATAC.K562.bam <span class="op">&gt;</span> ATAC.K562.bam.idxstats.txt</span></code></pre></div>
<p>Next, we count the DNase- or ATAC-seq counts along the genome, and save in Bigwig format. Having these reads counted allows us to efficiently extract the read counts around motif match positions for different motifs.</p>
<p><a href="https://github.com/kevinlkx/TOP/tree/main/inst/scripts/count_dnase_genome_cuts.R" class="external-link">Example script</a></p>
<p>We use the <code>bedtools</code> and <code>bedGraphToBigWig</code> and <code>bedSort</code> tools from UCSC.</p>
<p>Use the <code>fetchChromSizes</code> script to create the chrom.sizes file which will be needed.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fetchChromSizes</span> hg38 <span class="op">&gt;</span> hg38.chrom.sizes</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># bam_file: input BAM filename.</span>
<span class="co"># chrom_size_file: chrom sizes file.</span>
<span class="co"># outdir: Output directory.</span>
<span class="co"># outname: Output filename prefix.</span>
<span class="fu"><a href="../reference/count_dnase_genome_cuts.html">count_dnase_genome_cuts</a></span><span class="op">(</span>bam_file <span class="op">=</span> <span class="st">'ATAC.K562.bam'</span>, chrom_size_file <span class="op">=</span> <span class="st">'hg38.chrom.sizes'</span>, 
                        outdir <span class="op">=</span> <span class="st">"./processed_data/"</span>, outname <span class="op">=</span> <span class="st">"CTCF.K562.ATAC"</span><span class="op">)</span></code></pre></div>
<p><strong>Step 4: Get DNase- or ATAC-seq count matrices for each motif, then normalize, bin and transform the counts</strong></p>
<p>Get DNase or ATAC count matrices around candidate sites.</p>
<p><a href="https://github.com/kevinlkx/TOP/tree/main/inst/scripts/get_dnase_motif_counts.R" class="external-link">Example script</a></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># sites_file: Filename for candidate sites</span>
<span class="co"># dnase_fwd_count_file: Filename for DNase or ATAC counts in forward strand (Bigwig format)</span>
<span class="co"># dnase_rev_count_file: Filename for DNase or ATAC counts in reverse strand (Bigwig format)</span>
<span class="co"># dnase_fwd_matrix_file: Filename for DNase or ATAC count matrix in forward strand</span>
<span class="co"># dnase_rev_matrix_file: Filename for DNase or ATAC count matrix in reverse strand</span>
<span class="va">dnase_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_dnase_sites_counts.html">get_dnase_sites_counts</a></span><span class="op">(</span><span class="va">file_sites</span>, <span class="va">file_dnase_count_fwd</span>, <span class="va">file_dnase_count_rev</span>,
                                       <span class="va">file_dnase_matrix_fwd</span>, <span class="va">file_dnase_matrix_rev</span><span class="op">)</span></code></pre></div>
<p>Normalize, bin and transform DNase- or ATAC-seq counts.</p>
<p><a href="https://github.com/kevinlkx/TOP/tree/main/inst/scripts/normalize_bin_dnase.R" class="external-link">Example script</a></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># dnase_counts: DNase or ATAC count matrix</span>
<span class="co"># idxstats_file: The idxstats file generated by samtools</span>
<span class="co"># ref.size: Scale to reference library size </span>
<span class="co"># (Default: 100 million for DNase-seq and 50 million for ATAC-seq)</span>
<span class="co"># bin.method: MILLIPEDE binning scheme (Default: 'M5').</span>
<span class="co"># transform: asinh or log2 transform of DNase counts.</span>
<span class="va">dnase_bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize_bin_dnase.html">normalize_bin_dnase</a></span><span class="op">(</span><span class="va">dnase_counts</span>, <span class="va">idxstats_file</span>, 
                                  ref.size<span class="op">=</span><span class="fl">1e8</span>, bin.method<span class="op">=</span><span class="st">'M5'</span>, transform<span class="op">=</span><span class="st">'asinh'</span><span class="op">)</span></code></pre></div>
<p><strong>Step 5: Prepare ChIP-seq data if you want to train your own model (optional)</strong></p>
<p>Add a column with ChIP counts (from ChIP-seq reads) to the input data frame if you want to train the logistic version of the model.</p>
<p><a href="https://github.com/kevinlkx/TOP/tree/main/inst/scripts/count_chipseq_coverage.R" class="external-link">Example script</a></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pwm_id: Motif pwm ID</span>
<span class="co"># cell_type: Cell type</span>
<span class="co"># training_metadata: training metadata</span>
<span class="co"># sites_file: Filename for candidate sites</span>
<span class="co"># chrom_size_file: Filename for chromosome sizes</span>
<span class="co"># chip_dir: Directory of ChIP-seq bam files</span>
<span class="co"># outdir: Output directory</span>
<span class="co"># outname: Output filename prefix</span>
<span class="co"># ref.size: ChIP-Seq reference library size (Default: 10 million)</span>
<span class="va">chip_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_normalize_chip.html">count_normalize_chip</a></span><span class="op">(</span><span class="va">pwm_id</span>, <span class="va">cell_type</span>, <span class="va">training_metadata</span>,
                                    <span class="va">sites_file</span>, <span class="va">chrom_size_file</span>, <span class="va">chip_dir</span>, 
                                    <span class="va">outdir</span>, <span class="va">outname</span>, ref.size<span class="op">=</span><span class="fl">1e7</span><span class="op">)</span></code></pre></div>
<p>Add the binary ChIP labels (from ChIP-seq peaks) to the input data frame if you want to train the logistic version of the model</p>
<p>Finally, combine PWM scores, DNase (or ATAC) bins, and ChIP-seq counts or labels. <a href="https://github.com/kevinlkx/TOP/tree/main/inst/scripts/combine_dnase_chip_data.R" class="external-link">Example script</a></p>
</div>
<div class="section level2">
<h2 id="snakemake-pipeline">Snakemake pipeline<a class="anchor" aria-label="anchor" href="#snakemake-pipeline"></a>
</h2>
<p>We provided <a href="https://github.com/kevinlkx/TOP/tree/main/inst/snakemake" class="external-link">Snakemake pipelines</a> to automate the whole process using the above mentioned R scripts. The Snakemake is especially helpful if you have many TFs (motifs) in many cell types.</p>
<p>Run <code>Snakefile_training_ATAC</code> for ATAC-seq, or <code>Snakefile_training_DNase</code> for DNase-seq. For more details and instructions about running Snakemake pipelines, see <a href="https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html" class="external-link">Snakemake tutorial</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
