<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Prepare input data • TOP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Prepare input data">
<meta property="og:description" content="TOP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/installation.html">Install TOP R package</a>
    </li>
    <li>
      <a href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    </li>
    <li>
      <a href="../articles/prepare_input_data.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/train_TOP_model_JAGS.html">Train TOP model</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinlkx/TOP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Prepare input data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kevinlkx/TOP/blob/HEAD/vignettes/prepare_input_data.Rmd" class="external-link"><code>vignettes/prepare_input_data.Rmd</code></a></small>
      <div class="hidden name"><code>prepare_input_data.Rmd</code></div>

    </div>

    
    
<p>We provided scripts and a Snakemake pipeline to prepare candidate binding sites, DNase- or ATAC-seq bins from DNase- or ATAC-seq reads (bam files).</p>
<div class="section level2">
<h2 id="basic-procedure">Basic procedure<a class="anchor" aria-label="anchor" href="#basic-procedure"></a>
</h2>
<p>You can run the following steps as described here. You can use the <a href="https://github.com/HarteminkLab/TOP/tree/main/snakemake" class="external-link">Snakemake pipeline</a> below, if you have many TFs in many cell types and you want to train the whole model using all the input data from many TF x cell type combinations.</p>
<div class="section level3">
<h3 id="download-motif-database-and-install-required-softwares">Download motif database and install required softwares<a class="anchor" aria-label="anchor" href="#download-motif-database-and-install-required-softwares"></a>
</h3>
<ul>
<li><p>Download motifs from <a href="http://jaspar.genereg.net" class="external-link">JASPAR</a> motif database</p></li>
<li>
<p>Install softwares that will be needed later</p>
<ul>
<li>
<a href="http://meme-suite.org/doc/fimo.html?man_type=web" class="external-link">FIMO</a> from MEME suite.</li>
<li><a href="http://www.htslib.org" class="external-link">samtools</a></li>
<li><a href="http://bedtools.readthedocs.io/en/latest/" class="external-link">bedtools</a></li>
<li><a href="https://github.com/CRG-Barcelona/bwtool/wiki" class="external-link">bwtools</a></li>
<li>
<a href="http://hgdownload.cse.ucsc.edu/admin/exe/" class="external-link">UCSC command-line tools</a>: <code>bedGraphToBigWig</code>, <code>bigWigAverageOverBed</code>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="prepare-candidate-sites">Prepare candidate sites<a class="anchor" aria-label="anchor" href="#prepare-candidate-sites"></a>
</h3>
<div class="section level4">
<h4 id="find-tf-motif-matches-using-fimo">Find TF motif matches using <code>FIMO</code><a class="anchor" aria-label="anchor" href="#find-tf-motif-matches-using-fimo"></a>
</h4>
<p>To scan for TF motif matches, we recommend the <code>FIMO</code> software from the MEME suite:</p>
<ul>
<li><p>FIMO instructions: <a href="http://meme-suite.org/doc/fimo.html?man_type=web" class="external-link uri">http://meme-suite.org/doc/fimo.html?man_type=web</a></p></li>
<li><p>MEME Suite installation: <a href="http://meme-suite.org/doc/install.html?man_type=web" class="external-link uri">http://meme-suite.org/doc/install.html?man_type=web</a></p></li>
</ul>
<p>You will need the FIMO output (fimo.txt) in the following step.</p>
<p>You can follow the instructions from <a href="http://meme-suite.org/doc/fimo.html?man_type=web" class="external-link">FIMO</a>.</p>
<p>We use the command line version of FIMO. By default, we used with the threshold <span class="math inline">\(p &lt; 1e-5\)</span> and uniform background when training the model. You can choose your own background or use FIMO’s default threshold <span class="math inline">\(p &lt; 1e-4\)</span> (which will result in more motif matches).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example setting</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># outdir is the output directory</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># pwm: PWM file</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ref_genome: reference genome file</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>fimo <span class="sc">--</span>text <span class="sc">--</span>skip<span class="sc">-</span>matched<span class="sc">-</span>sequence <span class="sc">--</span>verbosity <span class="dv">2</span> \</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span>bgfile <span class="sc">--</span>uniform<span class="sc">--</span> \</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span>thresh <span class="fl">1e-5</span> <span class="sc">--</span>max<span class="sc">-</span>stored<span class="sc">-</span>scores <span class="dv">1000000</span> \</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="sc">--</span>oc outdir pwm ref_genome</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="get-candidate-tf-binding-sites-with-100bp-flanking-regions-around-motif-matches-obtained-from-fimo">Get candidate TF binding sites with 100bp flanking regions around motif matches obtained from FIMO<a class="anchor" aria-label="anchor" href="#get-candidate-tf-binding-sites-with-100bp-flanking-regions-around-motif-matches-obtained-from-fimo"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fimo: FIMO result filename. (fimo.txt)</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># blacklist: ENCODE blacklist file</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># flank: size (bp) of flanking region on each side of motif. Default: 100.</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># thresh_pValue: FIMO p-value threshold. </span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>Rscript ..<span class="sc">/</span>scripts<span class="sc">/</span>get_candidate_sites.R \</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>fimo fimo <span class="sc">--</span>blacklist blacklist \</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>flank <span class="dv">100</span> <span class="sc">--</span>thresh_pValue <span class="fl">1e-5</span> \</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>out output</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="prepare-dnase--or-atac-seq-data">Prepare DNase- or ATAC-seq data<a class="anchor" aria-label="anchor" href="#prepare-dnase--or-atac-seq-data"></a>
</h3>
<div class="section level4">
<h4 id="count-dnase--or-atac-seq-genome-wide-cleavage-and-output-tagcount-at-each-genome-position-in-bigwig-format">Count DNase- or ATAC-seq genome-wide cleavage, and output tagcount at each genome position in Bigwig format<a class="anchor" aria-label="anchor" href="#count-dnase--or-atac-seq-genome-wide-cleavage-and-output-tagcount-at-each-genome-position-in-bigwig-format"></a>
</h4>
<p>Sort and index the BAM file, and then retrieve and print stats in the index file.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>samtools index bam </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>samtools idxstats bam <span class="sc">&gt;</span> bam.idxstats</span></code></pre></div>
<p>Count the DNase- or ATAC-seq cleavage along the genome, output in Bigwig format</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># bam: input BAM filename.</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># chrom_size: filename of genome sizes by chromosomes.</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># out_dir: output directory</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># out_prefix: output prefix of the counts for genome-wide cuts</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>Rscript ..<span class="sc">/</span>scripts<span class="sc">/</span>count_dnase_genome_cuts.R \</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>bam bam \</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>chrom_size chrom_size \</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>out_dir genome_counts_dir \</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>out_prefix sample_name</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="match-dnase--or-atac-seq-cleavage-matrices-for-each-motifs-candidate-sites">Match DNase- or ATAC-seq cleavage matrices for each motif’s candidate sites<a class="anchor" aria-label="anchor" href="#match-dnase--or-atac-seq-cleavage-matrices-for-each-motifs-candidate-sites"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sites: filename for candidate sites</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># dnase_fwd: filename for DNase- or ATAC-seq genome cleavage counts in forward strand</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># dnase_rev: filename for DNase- or ATAC-seq genome cleavage counts in reverse strand</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># outdir: output directory</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># outname: output prefix of the counts for the motif</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>Rscript ..<span class="sc">/</span>scripts<span class="sc">/</span>get_dnase_motif_counts.R \</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>sites sites \</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>dnase_fwd fwd_genome_counts \</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>dnase_rev rev_genome_counts\</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>outdir outdir \</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>outname outname</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="normalize-bin-dnase--or-atac-seq-data-with-millipede-binning-scheme-and-tranform-asinh-dnase-counts">Normalize, bin DNase- or ATAC-seq data with MILLIPEDE binning scheme and tranform (asinh) DNase counts<a class="anchor" aria-label="anchor" href="#normalize-bin-dnase--or-atac-seq-data-with-millipede-binning-scheme-and-tranform-asinh-dnase-counts"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># dnase_counts: DNase- or ATAC-seq cleavage matrix</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># idxstats: idxstats file generated by samtools</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># bin: type of MILLIPEDE binning scheme (Default: 'M5').</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ref_size: normalize to reference library size (Default: 1e8 for DNase-seq and 5e7 for ATAC-seq)</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># transform: log2 or asinh transform</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>Rscript ..<span class="sc">/</span>scripts<span class="sc">/</span>normalize_bin_dnase.R \</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>dnase_counts dnase_counts \</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>idxstats idxstats \</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>bin <span class="st">'M5'</span> \</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>ref_size <span class="fl">1e8</span> \</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>transform <span class="st">'asinh'</span> \</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>outdir {outdir} \</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>outname {outname}</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="prepare-chip-seq-data-if-you-need-to-train-your-own-model">Prepare ChIP-seq data if you need to train your own model<a class="anchor" aria-label="anchor" href="#prepare-chip-seq-data-if-you-need-to-train-your-own-model"></a>
</h3>
<div class="section level4">
<h4 id="normalize-count-chip-seq-read-coverage">Normalize, count ChIP-seq read coverage<a class="anchor" aria-label="anchor" href="#normalize-count-chip-seq-read-coverage"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pwm_id: motif pwm ID.</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># cell_type: cell type.</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># sites: Filename for candidate sites.</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># chrom_size: Filename listing the size for each chromosome.</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># metadata: Filename for training metadata table.</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># chip_dir: Directory for ChIP-seq bam files.</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ref_size: Reference library size for ChIP-seq. (Default: 2e7)</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># outdir: Output directory.</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co"># outname: Output filename prefix.</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>Rscript ..<span class="sc">/</span>scripts<span class="sc">/</span>count_chipseq_coverage.R \</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>pwm_id pwm_id \</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>cell_type cell_type \</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>sites sites \</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>chip_dir chip_dir \</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>chrom_size chrom_size \</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>metadata chrom_size \</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>ref_size <span class="fl">2e7</span> \</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>outdir chip_counts_dir \</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>outname outname</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="combine-pwm-scores-dnase-or-atac-bins-and-chip-seq-counts-for-candidate-sites">Combine PWM scores, DNase (or ATAC) bins, and ChIP-seq counts for candidate sites<a class="anchor" aria-label="anchor" href="#combine-pwm-scores-dnase-or-atac-bins-and-chip-seq-counts-for-candidate-sites"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tf_name: TF name.</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># cell_type: cell type.</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># metadata: Filename for training metadata table.</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># thresh_pValue: FIMO p-value threshold. </span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># bin: type of MILLIPEDE binning scheme (Default: 'M5').</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co"># chip_counts: ChIP read counts.</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># dnase_dir: Directory for DNase or ATAC count matrices.</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># dnase_idxstats_dir: Directory for DNase or ATAC reads idxstats files.</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># outdir: Output directory.</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="co"># outname: Output filename prefix.</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>Rscript ..<span class="sc">/</span>scripts<span class="sc">/</span>combine_dnase_chip_data.R \</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>tf_name tf_name \</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>cell_type cell_type \</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>metadata metadata \</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>thresh_pValue <span class="fl">1e-5</span> \</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>bin <span class="st">'M5'</span> \</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>chip_counts chip_counts \</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>dnase_dir dnase_dir \</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>dnase_idxstats_dir dnase_idxstats_dir \</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>outdir outdir \</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>outname outname</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="add-chip-binary-labels-from-chip-seq-peaks-to-the-combined-training-data">Add ChIP binary labels (from ChIP-seq peaks) to the combined training data<a class="anchor" aria-label="anchor" href="#add-chip-binary-labels-from-chip-seq-peaks-to-the-combined-training-data"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tf_name: TF name.</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co"># cell_type: cell type.</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># metadata: Filename for training metadata table.</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># thresh_pValue: FIMO p-value threshold. </span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># bin: type of MILLIPEDE binning scheme (Default: 'M5').</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co"># combined_data_dir: Directory for combined data for each TF x cell type combo.</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co"># chip_peak_dir: Directory for ChIP-seq peak bed narrowPeak files.</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>Rscript ..<span class="sc">/</span>scripts<span class="sc">/</span>add_chip_peaks_to_combined_data.R \</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>tf_name tf_name \</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>cell_type cell_type \</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>metadata metadata \</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>thresh_pValue <span class="fl">1e-5</span> \</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>bin <span class="st">'M5'</span> \</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>combined_data_dir combined_data_dir \</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>        <span class="sc">--</span>chip_peak_dir chip_peak_dir</span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="snakemake-pipeline">Snakemake pipeline<a class="anchor" aria-label="anchor" href="#snakemake-pipeline"></a>
</h2>
<p>We provided a <a href="https://github.com/HarteminkLab/TOP/tree/main/snakemake" class="external-link">Snakemake pipeline</a> to automate the whole process and helpful if you have many TFs (motifs) in many cell types.</p>
<p>Run <code>Snakefile_training_ATAC</code> for ATAC-seq, or <code>Snakefile_training_DNase</code> for DNase-seq. For more details and instructions about running Snakemake pipelines, see <a href="https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html" class="external-link">this page</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
