<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Train TOP logistic model • TOP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Train TOP logistic model">
<meta property="og:description" content="TOP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_model.html">Train TOP quantitative occupancy model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_logistic_model.html">Train TOP logistic (binary binding) model</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinlkx/TOP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Train TOP logistic model</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kevinlkx/TOP/blob/HEAD/vignettes/train_TOP_logistic_model.Rmd" class="external-link"><code>vignettes/train_TOP_logistic_model.Rmd</code></a></small>
      <div class="hidden name"><code>train_TOP_logistic_model.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="prepare-training-data">Prepare training data<a class="anchor" aria-label="anchor" href="#prepare-training-data"></a>
</h2>
<p>Similar to the steps in <a href="train_TOP_model.html">Training TOP quantitative occupancy model</a>, We need to prepare and assemble the training data containing all training TF x cell type combinations.</p>
<p><strong>Step 1: Prepare training data for each TF in each cell type</strong></p>
<p>Firstly, we need to prepare training data for each training TF x cell type combination, and save the training data files (as <code>.rds</code> files in the <code>data_file</code> column in the table below).</p>
<p>For each TF in each cell type, we prepare training data for candidate binding sites: including:</p>
<ul>
<li>PWM scores of the motif matches</li>
<li>DNase (or ATAC) counts in MILLIPEDE bins</li>
<li>
<em>binary ChIP labels</em> (from ChIP-seq peaks).</li>
</ul>
<p>You can follow this <a href="data_preparation.html">procedure</a> to prepare the training data.</p>
<p><strong>Step 2: Assemble training data for all TF-cell type combinations</strong></p>
<p>We create a table (data frame) listing all training TF x cell type combinations. The table should have three columns: TF names, cell types, and paths to the training data files, like:</p>
<table class="table">
<thead><tr class="header">
<th align="center">tf_name</th>
<th align="center">cell_type</th>
<th align="center">data_file</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">CTCF</td>
<td align="center">K562</td>
<td align="center">CTCF.K562.data.rds</td>
</tr>
<tr class="even">
<td align="center">CTCF</td>
<td align="center">A549</td>
<td align="center">CTCF.A549.data.rds</td>
</tr>
<tr class="odd">
<td align="center">CTCF</td>
<td align="center">GM12878</td>
<td align="center">CTCF.GM12878.data.rds</td>
</tr>
<tr class="even">
<td align="center">NRF1</td>
<td align="center">K562</td>
<td align="center">NRF1.K562.data.rds</td>
</tr>
<tr class="odd">
<td align="center">MYC</td>
<td align="center">K562</td>
<td align="center">MYC.K562.data.rds</td>
</tr>
<tr class="even">
<td align="center">…</td>
<td align="center">…</td>
<td align="center">…</td>
</tr>
</tbody>
</table>
<p>Use the following <code><a href="../reference/assemble_training_data.html">assemble_training_data()</a></code> function, we split the training data randomly into 10 equal partitions, so that we could run Gibbs sampling on these partitions in parallel to reduce the running time.</p>
<p>We can choose the training chromosomes by specifying the chromosomes in <code>training_chrs</code>, for example using odd chromosomes as below. The chromosome names (“chr…”) should match with those in the training data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinlkx/TOP" class="external-link">TOP</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">assembled_training_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assemble_training_data.html">assemble_training_data</a></span><span class="op">(</span><span class="va">tf_cell_table</span>,
                                                  logistic.model <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># use logistic model</span>
                                                  chip_colname <span class="op">=</span> <span class="st">'chip_label'</span>, <span class="co"># name of the column with ChIP labels in training data</span>
                                                  training_chrs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'chr'</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">21</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>, 
                                                  n.partitions <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p>Save a table listing all TF and cell type combinations. This will be used later to extract regression coefficients.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tf_cell_combos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">assembled_training_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'tf_id'</span>, <span class="st">'cell_id'</span>, <span class="st">'tf_name'</span>, <span class="st">'cell_type'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fit-top-logistic-models-using-assembled-training-data">Fit TOP logistic models using assembled training data<a class="anchor" aria-label="anchor" href="#fit-top-logistic-models-using-assembled-training-data"></a>
</h2>
<p>Here, we fit TOP logistic models with ChIP-seq binary labels, to fit TOP quantitative occupancy models with ChIP-seq read counts, please follow this <a href="train_TOP_model.htmll">tutorial</a>.</p>
<p>We fit TOP models using <a href="https://mcmc-jags.sourceforge.io" class="external-link">JAGS</a>, and we uses the <a href="https://cran.r-project.org/web/packages/R2jags/" class="external-link">R2jags</a> R package to call JAGS from R.</p>
<p>Please install <code>R2jags</code> if you need to train your own TOP models.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">R2jags</span><span class="op">)</span></code></pre></div>
<p>We provided the TOP model files in JAGS code in the <code>../model</code> directory within the package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model_file</span> <span class="op">&lt;-</span> <span class="st">'../model/TOP_M5_logit_model.jags'</span></code></pre></div>
<p>The <code>fit_TOP_model()</code> function below runs Gibbs sampling for each of the 10 partitions in parallel (using the <code>foreach</code> R package).</p>
<p>We can set the following parameters for Gibbs sampling:</p>
<ul>
<li>n.iter: number of total iterations per chain (including burn-in iterations).</li>
<li>n.burnin: number of burn-in iterations, i.e. number of iterations to discard at the beginning.</li>
<li>n.chains: number of Markov chains.</li>
<li>n.thin: thinning rate, must be a positive integer.</li>
</ul>
<p>The following example runs 10000 iterations of Gibbs sampling in total, with 2000 burn-ins, 3 Markov chains, thinning rate of 2, and save the posterior samples to the <code>TOP_logistic_fit</code> directory.</p>
<p>It could take a long time if we have many TFs and many cell types in the training data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TOP_samples_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_TOP_M5_model.html">fit_TOP_M5_model</a></span><span class="op">(</span><span class="va">assembled_training_data</span>, 
                                      model.file <span class="op">=</span> <span class="va">model_file</span>,
                                      logistic.model <span class="op">=</span> <span class="cn">TRUE</span>,
                                      out.dir <span class="op">=</span> <span class="st">'TOP_logistic_fit'</span>,
                                      partitions <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,
                                      n.iter <span class="op">=</span> <span class="fl">5000</span>, 
                                      n.burnin <span class="op">=</span> <span class="fl">2000</span>,
                                      n.chains <span class="op">=</span> <span class="fl">3</span>,
                                      n.thin <span class="op">=</span> <span class="fl">2</span>,
                                      quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Alternatively, we can also submit jobs for the partitions (choose which partition to run, e.g. <code>partitions=1</code>) in parallel to separate compute nodes on a cluster.</p>
</div>
<div class="section level2">
<h2 id="combine-top-posterior-samples-from-all-partitions-and-extract-the-posterior-mean-of-the-regression-coefficients">Combine TOP posterior samples from all partitions and extract the posterior mean of the regression coefficients<a class="anchor" aria-label="anchor" href="#combine-top-posterior-samples-from-all-partitions-and-extract-the-posterior-mean-of-the-regression-coefficients"></a>
</h2>
<p>After we finished Gibbs sampling for all 10 partitions, we combine the posterior samples from all the partitions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TOP_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_TOP_samples.html">combine_TOP_samples</a></span><span class="op">(</span><span class="va">TOP_samples_files</span><span class="op">)</span></code></pre></div>
<p>We can extract posterior mean of the coefficients for all three levels.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TOP_mean_coef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_TOP_mean_coef.html">extract_TOP_mean_coef</a></span><span class="op">(</span><span class="va">TOP_samples</span>, <span class="va">tf_cell_combos</span><span class="op">)</span></code></pre></div>
<p>Save posterior samples and posterior mean of the regression coefficients.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">TOP_samples</span>, <span class="st">'TOP_logistic_fit/TOP_logistic_M5_combined_posterior_samples.rds'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">TOP_mean_coef</span>, <span class="st">'TOP_logistic_fit/TOP_logistic_M5_posterior_mean_coef.rds'</span><span class="op">)</span></code></pre></div>
<p>We will provide pre-trained models using ENCODE data <a href="https://users.cs.duke.edu/~amink/software/" class="external-link">here</a>.</p>
<p>To make predictions for TF occupancy using new DNase- or ATAC-seq data using trained model coefficients, see this <a href="predict_TF_occupancy_with_trained_model.html">page</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
