<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Train TOP quantitative occupancy model • TOP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Train TOP quantitative occupancy model">
<meta property="og:description" content="TOP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.13.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_model.html">Train TOP quantitative occupancy model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_logistic_model.html">Train TOP logistic (binary binding) model</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HarteminkLab/TOP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Train TOP quantitative occupancy model</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/HarteminkLab/TOP/blob/HEAD/vignettes/train_TOP_model.Rmd" class="external-link"><code>vignettes/train_TOP_model.Rmd</code></a></small>
      <div class="hidden name"><code>train_TOP_model.Rmd</code></div>

    </div>

    
    
<p>TOP jointly fits a Bayesian hierarchical model using available training data from many TFs in many cell types.</p>
<p>Training TOP entailed two basic steps, as illustrated in the figure below:</p>
<p><img src="../reference/figures/schematic.png" width="90%"></p>
<p>First, following the site-centric strategy, we used motif matches to enumerate candidate binding sites, and extracted DNase- and/or ATAC-seq and ChIP data centered on each site for training.</p>
<p>Second, we fit our Bayesian hierarchical regression model on spatially-binned DNase- and/or ATAC-seq data.</p>
<p>Once TOP is trained, we can use it to predict occupancy for cell types or conditions for which we have DNase- or ATAC-seq data, generating TF occupancy estimates for ChIP-seq experiments that have not been done.</p>
<div class="section level2">
<h2 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<p>We need to have R packages <a href="https://cran.r-project.org/web/packages/R2jags/index.html" class="external-link">R2jags</a> and <a href="https://cran.r-project.org/web/packages/doParallel/index.html" class="external-link">doParallel</a> installed to train TOP models, as we use <a href="https://mcmc-jags.sourceforge.io" class="external-link">JAGS</a> to run Gibbs sampling for the Bayesian hierarchical model in the current implementation of the model.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HarteminkLab/TOP" class="external-link">TOP</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prepare-training-data">Prepare training data<a class="anchor" aria-label="anchor" href="#prepare-training-data"></a>
</h2>
<p>As we learn the parameters from many TFs in many cell types together, it takes some efforts to prepare the training data.</p>
<p><strong>Step 1: Prepare training data for each TF in each cell type</strong></p>
<p>Firstly, we need to prepare training data for each training TF x cell type combination, and save the training data files (as <code>.rds</code> files in the <code>data_file</code> column in the table below).</p>
<p>This <a href="data_preparation.html">page</a> shows the procedure to prepare the training data. Basically, for each TF in each cell type, we prepare training data for candidate binding sites: including: PWM scores of the motif matches, DNase (or ATAC) counts in 5 bins, and measured TF occupancy (from ChIP-seq data).</p>
<p><strong>Step 2: Assemble training data for all TF x cell type combinations</strong></p>
<p>We create a table (data frame) listing all training TF x cell type combinations. The table should have three columns: TF names, cell types, and paths to the training data files, like:</p>
<table class="table">
<thead><tr class="header">
<th align="center">tf_name</th>
<th align="center">cell_type</th>
<th align="center">data_file</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">CTCF</td>
<td align="center">K562</td>
<td align="center">CTCF.K562.data.rds</td>
</tr>
<tr class="even">
<td align="center">CTCF</td>
<td align="center">A549</td>
<td align="center">CTCF.A549.data.rds</td>
</tr>
<tr class="odd">
<td align="center">CTCF</td>
<td align="center">GM12878</td>
<td align="center">CTCF.GM12878.data.rds</td>
</tr>
<tr class="even">
<td align="center">NRF1</td>
<td align="center">K562</td>
<td align="center">NRF1.K562.data.rds</td>
</tr>
<tr class="odd">
<td align="center">MYC</td>
<td align="center">K562</td>
<td align="center">MYC.K562.data.rds</td>
</tr>
<tr class="even">
<td align="center">…</td>
<td align="center">…</td>
<td align="center">…</td>
</tr>
</tbody>
</table>
<p>In the example below, we split the training data randomly into 10 equal partitions, so that we could run Gibbs sampling on these partitions separately in parallel to reduce the running time.</p>
<p>We choose the training chromosomes by specifying the chromosomes in <code>training_chrs</code>, for example using odd chromosomes as below. The chromosome names (“chr…”) should match with those in the training data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">assembled_training_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assemble_training_data.html">assemble_training_data</a></span><span class="op">(</span><span class="va">tf_cell_table</span>,
                                                  training_chrs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'chr'</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">21</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>, 
                                                  n_partitions <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fit-top-quantitative-occupancy-models-using-assembled-training-data">Fit TOP quantitative occupancy models using assembled training data<a class="anchor" aria-label="anchor" href="#fit-top-quantitative-occupancy-models-using-assembled-training-data"></a>
</h2>
<p>Here, we fit TOP quantitative occupancy models with ChIP-seq read counts. To fit TOP logistic models with ChIP-seq binary labels, please follow this <a href="train_TOP_logistic_model.htmll">tutorial</a>.</p>
<p>We use the consensus Monte Carlo algorithm to run Gibbs sampling on each of the 10 training data partitions separately in parallel.</p>
<p>The <code>fit_TOP_model()</code> function below runs Gibbs sampling for each of the 10 partitions in parallel.</p>
<p>We first perform “asinh” transform to the training ChIP-seq counts, then fit the TOP quantitative occupancy model for each of the 10 partitions in parallel.</p>
<p>Set the following parameters for Gibbs sampling:</p>
<ul>
<li>n_iter: number of total iterations per chain (including burn-in iterations).</li>
<li>n_burnin: number of burn-in iterations, i.e. number of iterations to discard at the beginning.</li>
<li>n_chains: number of Markov chains.</li>
<li>n_thin: thinning rate, must be a positive integer.</li>
</ul>
<p>The following example runs 5000 iterations of Gibbs sampling in total, including 1000 burn-ins, with 3 Markov chains, at a thinning rate of 2, and save the posterior samples to the <code>TOP_fit</code> directory.</p>
<p>It could take a long time if we have many TFs and many cell types in the training data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">all_TOP_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_TOP_M5_model.html">fit_TOP_M5_model</a></span><span class="op">(</span><span class="va">assembled_training_data</span>, 
                                    logistic_model <span class="op">=</span> <span class="cn">FALSE</span>,
                                    transform <span class="op">=</span> <span class="st">'asinh'</span>, 
                                    n_iter <span class="op">=</span> <span class="fl">5000</span>,
                                    n_burnin <span class="op">=</span> <span class="fl">1000</span>,
                                    n_chains <span class="op">=</span> <span class="fl">3</span>,
                                    n_thin <span class="op">=</span> <span class="fl">2</span>,
                                    out_dir <span class="op">=</span> <span class="st">'TOP_fit'</span><span class="op">)</span></code></pre></div>
<p>This fits a TOP model on all 10 partitions in parallel on 10 CPU cores, and returns a list of posterior samples of the coefficients for each of the 10 partitions. It requires a compute node/machine with 10 cores and may require a big memory if you have training data from many TFs and cell types. Alternatively, if you have limited computing resource, you may fit model for each of the 10 the partitions on separate machines, by specifying the partition to run. For example, setting argument <code>partitions=3</code> will fit TOP model for training data in the 3rd partition.</p>
</div>
<div class="section level2">
<h2 id="combine-top-posterior-samples-from-all-partitions-and-extract-the-posterior-mean-of-the-regression-coefficients">Combine TOP posterior samples from all partitions and extract the posterior mean of the regression coefficients<a class="anchor" aria-label="anchor" href="#combine-top-posterior-samples-from-all-partitions-and-extract-the-posterior-mean-of-the-regression-coefficients"></a>
</h2>
<p>After we finished Gibbs sampling for all 10 partitions, we select and combine the posterior samples from all the partitions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TOP_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_TOP_samples.html">combine_TOP_samples</a></span><span class="op">(</span><span class="va">all_TOP_samples</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">TOP_samples</span><span class="op">)</span></code></pre></div>
<p>We can extract posterior mean of the coefficients for all three levels.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TOP_mean_coef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_TOP_mean_coef.html">extract_TOP_mean_coef</a></span><span class="op">(</span><span class="va">TOP_samples</span>, 
                         assembled_training_data <span class="op">=</span> <span class="va">assembled_training_data</span><span class="op">)</span></code></pre></div>
<p>Save the posterior samples and posterior mean coefficients for predictions.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">TOP_samples</span>, <span class="st">'TOP_fit/TOP_M5_combined_posterior_samples.rds'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">TOP_mean_coef</span>, <span class="st">'TOP_fit/TOP_M5_posterior_mean_coef.rds'</span><span class="op">)</span></code></pre></div>
<p>We will provide pre-trained models using ENCODE data <a href="https://users.cs.duke.edu/~amink/software/" class="external-link">here</a>.</p>
<p>To make predictions for TF occupancy using new DNase- or ATAC-seq data using trained model coefficients, please follow this tutorial <a href="predict_TF_occupancy_with_trained_model.html">“Predict TF occupancy using trained TOP model”</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
