<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Counts DNase-seq or ATAC-seq cuts along the genome — count_genome_cuts • TOP</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Counts DNase-seq or ATAC-seq cuts along the genome — count_genome_cuts"><meta property="og:description" content="Counts genomic cuts (5' end) from DNase-seq or
ATAC-seq BAM alignment files using bedtools
For ATAC-seq, when shift_ATAC = TRUE, shifts reads
so as to address offsets and align the signal across strands."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.12.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_preparation.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_model.html">Train TOP quantitative occupancy model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_logistic_model.html">Train TOP logistic (binary binding) model</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kevinlkx/TOP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Counts DNase-seq or ATAC-seq cuts along the genome</h1>
    <small class="dont-index">Source: <a href="https://github.com/kevinlkx/TOP/blob/HEAD/R/process_dnase_atac_data.R" class="external-link"><code>R/process_dnase_atac_data.R</code></a></small>
    <div class="hidden name"><code>count_genome_cuts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Counts genomic cuts (5' end) from DNase-seq or
ATAC-seq BAM alignment files using <code>bedtools</code>
For ATAC-seq, when <code>shift_ATAC = TRUE</code>, shifts reads
so as to address offsets and align the signal across strands.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">count_genome_cuts</span><span class="op">(</span>
  <span class="va">bam_file</span>,
  <span class="va">chrom_size_file</span>,
  data_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DNase"</span>, <span class="st">"ATAC"</span><span class="op">)</span>,
  shift_ATAC <span class="op">=</span> <span class="cn">TRUE</span>,
  shift_ATAC_bases <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4L</span>, <span class="op">-</span><span class="fl">4L</span><span class="op">)</span>,
  outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">dirname</a></span><span class="op">(</span><span class="va">bam_file</span><span class="op">)</span>,
  <span class="va">outname</span>,
  bedtools_path <span class="op">=</span> <span class="st">"bedtools"</span>,
  bedGraphToBigWig_path <span class="op">=</span> <span class="st">"bedGraphToBigWig"</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>bam_file</dt>
<dd><p>Sorted BAM file.</p></dd>
<dt>chrom_size_file</dt>
<dd><p>Chromosome size file.</p></dd>
<dt>data_type</dt>
<dd><p>Data type. Options: ‘DNase’ or ‘ATAC’.</p></dd>
<dt>shift_ATAC</dt>
<dd><p>Logical. When <code>shift_ATAC=TRUE</code> (and <code>data_type='ATAC'</code>),
shifts reads according to <code>shift_ATAC_bases</code>.</p></dd>
<dt>shift_ATAC_bases</dt>
<dd><p>Number of bases to shift on + and - strands.
Default: shifts reads on + strand by 4 bp and reads on - strand by -4 bp.</p></dd>
<dt>outdir</dt>
<dd><p>Output directory (default: use the directory of <code>bam_file</code>).</p></dd>
<dt>outname</dt>
<dd><p>Output prefix (default: use the prefix of <code>bam_file</code>).</p></dd>
<dt>bedtools_path</dt>
<dd><p>Path to <code>bedtools</code> executable.</p></dd>
<dt>bedGraphToBigWig_path</dt>
<dd><p>Path to UCSC <code>bedGraphToBigWig</code> executable.</p></dd>
</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="co"># ATAC-seq data</span></span>
<span class="r-in"><span class="fu">count_genome_cuts</span><span class="op">(</span>bam_file<span class="op">=</span><span class="st">'K562.ATAC.bam'</span>,</span>
<span class="r-in">                  chrom_size_file<span class="op">=</span><span class="st">'hg38.chrom.sizes'</span>,</span>
<span class="r-in">                  data_type<span class="op">=</span><span class="st">'ATAC'</span>,</span>
<span class="r-in">                  shift_ATAC<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span class="r-in">                  outdir<span class="op">=</span><span class="st">'processed_data'</span>,</span>
<span class="r-in">                  outname<span class="op">=</span><span class="st">'K562.ATAC'</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># DNase-seq data</span></span>
<span class="r-in"><span class="fu">count_genome_cuts</span><span class="op">(</span>bam_file<span class="op">=</span><span class="st">'K562.DNase.bam'</span>,</span>
<span class="r-in">                  chrom_size_file<span class="op">=</span><span class="st">'hg38.chrom.sizes'</span>,</span>
<span class="r-in">                  data_type<span class="op">=</span><span class="st">'DNase'</span>,</span>
<span class="r-in">                  outdir<span class="op">=</span><span class="st">'processed_data'</span>,</span>
<span class="r-in">                  outname<span class="op">=</span><span class="st">'K562.DNase'</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

