<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get count matrices around candidate binding sites — get_sites_counts • TOP</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get count matrices around candidate binding sites — get_sites_counts"><meta property="og:description" content="Extract counts around candidate binding sites on both strands
from the genome counts (BigWig files prepared using count_genome_cuts).
It utilizes the extract bed function from the bwtool software
to extract the read counts.
Then combine the counts into one matrix, with the first half of the columns
representing the read counts on the forward strand,
and the second half of the columns representing the read counts
on the reverse strand."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_preparation.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_model.html">Train TOP quantitative occupancy model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_logistic_model.html">Train TOP logistic (binary binding) model</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kevinlkx/TOP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get count matrices around candidate binding sites</h1>
    <small class="dont-index">Source: <a href="https://github.com/kevinlkx/TOP/blob/HEAD/R/dnase_atac_utils.R" class="external-link"><code>R/dnase_atac_utils.R</code></a></small>
    <div class="hidden name"><code>get_sites_counts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Extract counts around candidate binding sites on both strands
from the genome counts (BigWig files prepared using <code>count_genome_cuts</code>).
It utilizes the <code>extract bed</code> function from the <code>bwtool</code> software
to extract the read counts.
Then combine the counts into one matrix, with the first half of the columns
representing the read counts on the forward strand,
and the second half of the columns representing the read counts
on the reverse strand.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">get_sites_counts</span><span class="op">(</span>
  <span class="va">sites</span>,
  <span class="va">genomecount_dir</span>,
  <span class="va">genomecount_name</span>,
  tmpdir <span class="op">=</span> <span class="va">genomecount_dir</span>,
  bwtool_path <span class="op">=</span> <span class="st">"bwtool"</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sites</dt>
<dd><p>A data frame containing the candidate sites.</p></dd>
<dt>genomecount_dir</dt>
<dd><p>Directory for genome counts.</p></dd>
<dt>genomecount_name</dt>
<dd><p>File prefix for genome counts.</p></dd>
<dt>tmpdir</dt>
<dd><p>Temporary directory to save intermediate files.</p></dd>
<dt>bwtool_path</dt>
<dd><p>Path to <code>bwtool</code> executable.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A count matrix. The first half of the columns
are the read counts on the forward strand, and the second half of the
columns are the read counts on the reverse strand.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"><span class="co"># Example: Obtain the ATAC-seq count matrices around</span></span>
<span class="r-in"><span class="co"># candidate binding sites of CTCF (using JASPAR motif: MA0139.1)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># 1. Find TF motif matches using FIMO</span></span>
<span class="r-in"><span class="fu"><a href="fimo_motif_matches.html">fimo_motif_matches</a></span><span class="op">(</span>motif_file<span class="op">=</span><span class="st">'MA0139.1.meme'</span>, sequence_file<span class="op">=</span><span class="st">'hg38.fa'</span>,</span>
<span class="r-in">                   thresh_pValue<span class="op">=</span><span class="fl">1e-5</span>, max_strand<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span class="r-in">                   outname<span class="op">=</span><span class="st">'MA0139.1_1e-5.fimo.txt'</span>,</span>
<span class="r-in">                   fimo_path<span class="op">=</span><span class="st">'fimo'</span><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in fimo_motif_matches(motif_file = "MA0139.1.meme", sequence_file = "hg38.fa",     thresh_pValue = 1e-05, max_strand = TRUE, outname = "MA0139.1_1e-5.fimo.txt",     fimo_path = "fimo"):</span> fimo could not be executed. Please install fimo and set fimo_path.</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># 2. Get candidate TF binding sites</span></span>
<span class="r-in"><span class="va">sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="process_candidate_sites.html">process_candidate_sites</a></span><span class="op">(</span>fimo_file<span class="op">=</span><span class="st">'MA0139.1_1e-5.fimo.txt'</span>,</span>
<span class="r-in">                                 thresh_pValue<span class="op">=</span><span class="fl">1e-5</span>,</span>
<span class="r-in">                                 blacklist_file<span class="op">=</span><span class="st">'blacklist.hg38.bed.gz'</span><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in process_candidate_sites(fimo_file = "MA0139.1_1e-5.fimo.txt",     thresh_pValue = 1e-05, blacklist_file = "blacklist.hg38.bed.gz"):</span> MA0139.1_1e-5.fimo.txt file does not exist or is empty!</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># 3. Get ATAC-seq counts along the genome</span></span>
<span class="r-in"><span class="co"># Download the BAM file, (sort if the BAM file is unsorted),</span></span>
<span class="r-in"><span class="co"># index the BAM file, and</span></span>
<span class="r-in"><span class="co"># obtain the total number of mapped reads from the idxstats file.</span></span>
<span class="r-in"><span class="fu"><a href="sort_index_idxstats_bam.html">sort_index_idxstats_bam</a></span><span class="op">(</span><span class="st">'K562.ATAC.bam'</span>, sort<span class="op">=</span><span class="cn">FALSE</span>, index<span class="op">=</span><span class="cn">TRUE</span>, idxstats<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Index BAM file...</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in FUN(X[[i]], ...):</span> failed to open SAM/BAM file</span>
<span class="r-err co"><span class="r-pr">#&gt;</span>   file: 'K562.ATAC.bam'</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Obtain ATAC-seq counts along the genome</span></span>
<span class="r-in"><span class="fu"><a href="count_genome_cuts.html">count_genome_cuts</a></span><span class="op">(</span>bam_file<span class="op">=</span><span class="st">'K562.ATAC.bam'</span>,</span>
<span class="r-in">                  chrom_size_file<span class="op">=</span><span class="st">'hg38.chrom.sizes'</span>,</span>
<span class="r-in">                  data_type<span class="op">=</span><span class="st">'ATAC'</span>,</span>
<span class="r-in">                  outdir<span class="op">=</span><span class="st">'processed_data'</span>,</span>
<span class="r-in">                  outname<span class="op">=</span><span class="st">'K562.ATAC'</span>,</span>
<span class="r-in">                  bedtools_path<span class="op">=</span><span class="st">'bedtools'</span>,</span>
<span class="r-in">                  bedGraphToBigWig_path<span class="op">=</span><span class="st">'bedGraphToBigWig'</span><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in count_genome_cuts(bam_file = "K562.ATAC.bam", chrom_size_file = "hg38.chrom.sizes",     data_type = "ATAC", outdir = "processed_data", outname = "K562.ATAC",     bedtools_path = "bedtools", bedGraphToBigWig_path = "bedGraphToBigWig"):</span> bedtools could not be executed. Please install bedtools and set bedtools_path.</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># 4. Get ATAC-seq count matrices around candidate sites,</span></span>
<span class="r-in"><span class="co"># then normalize, bin and transform the counts</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Get ATAC-seq count matrices</span></span>
<span class="r-in"><span class="va">sites_counts.mat</span> <span class="op">&lt;-</span> <span class="fu">get_sites_counts</span><span class="op">(</span><span class="va">sites</span>,</span>
<span class="r-in">                                     genomecount_dir<span class="op">=</span><span class="st">'processed_data'</span>,</span>
<span class="r-in">                                     genomecount_name<span class="op">=</span><span class="st">'K562.ATAC'</span>,</span>
<span class="r-in">                                     bwtool_path<span class="op">=</span><span class="st">'bwtool'</span><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in get_sites_counts(sites, genomecount_dir = "processed_data", genomecount_name = "K562.ATAC",     bwtool_path = "bwtool"):</span> bwtool could not be executed. Please install bwtool and set bwtool_path.</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Normalize, bin and transform the counts</span></span>
<span class="r-in"><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize_bin_transform_counts.html">normalize_bin_transform_counts</a></span><span class="op">(</span><span class="va">sites_counts.mat</span>,</span>
<span class="r-in">                                       idxstats_file<span class="op">=</span><span class="st">'K562.ATAC.bam.idxstats.txt'</span>,</span>
<span class="r-in">                                       ref_size<span class="op">=</span><span class="fl">5e7</span>,</span>
<span class="r-in">                                       bin_method<span class="op">=</span><span class="st">'M5'</span>,</span>
<span class="r-in">                                       transform<span class="op">=</span><span class="st">'asinh'</span><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in data.table::fread(idxstats_file, sep = "\t"):</span> File 'K562.ATAC.bam.idxstats.txt' does not exist or is non-readable. getwd()=='/Users/kaixuanluo/Dropbox/projects_dropbox/TOP/R_package/TOP/docs/reference'</span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

