<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ TOP</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">TOP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/data_preparation.html">Prepare input data</a>
    <a class="dropdown-item" href="../articles/installation.html">Install TOP R package</a>
    <a class="dropdown-item" href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    <a class="dropdown-item" href="../articles/train_TOP_model_JAGS.html">Train TOP model</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/kevinlkx/TOP/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="assemble_TOP_logistic_training_data.html">assemble_TOP_logistic_training_data()</a></code> 
        </dt>
        <dd>Assemble TOP (logistic version) training data for all TF x cell type combos,
then split training data into partitions
#'</dd>
      </dl><dl><dt>
          
          <code><a href="assemble_TOP_training_data.html">assemble_TOP_training_data()</a></code> 
        </dt>
        <dd>Assemble TOP training data for all TF x cell type combos,
then split training data into partitions</dd>
      </dl><dl><dt>
          
          <code><a href="bam_sort_index_stats.html">bam_sort_index_stats()</a></code> 
        </dt>
        <dd>Sort and index the BAM file, and then retrieve and print stats in the index file.</dd>
      </dl><dl><dt>
          
          <code><a href="bin_dnase.html">bin_dnase()</a></code> 
        </dt>
        <dd>Binning and transform DNase count matrix</dd>
      </dl><dl><dt>
          
          <code><a href="combine_TOP_samples.html">combine_TOP_samples()</a></code> 
        </dt>
        <dd>Combine and take the average of TOP posterior samples from all partitions</dd>
      </dl><dl><dt>
          
          <code><a href="combine_dnase_bins_chip_data.html">combine_dnase_bins_chip_data()</a></code> 
        </dt>
        <dd>Combined normalized and binned DNase data with normalized ChIP counts data</dd>
      </dl><dl><dt>
          
          <code><a href="compute_mapability.html">compute_mapability()</a></code> 
        </dt>
        <dd>Title</dd>
      </dl><dl><dt>
          
          <code><a href="count_dnase_genome_cuts.html">count_dnase_genome_cuts()</a></code> 
        </dt>
        <dd>Count DNase-seq or ATAC-seq cleavage for a given genome.</dd>
      </dl><dl><dt>
          
          <code><a href="count_normalize_chip.html">count_normalize_chip()</a></code> 
        </dt>
        <dd>Count and normalize ChIP-seq read coverage</dd>
      </dl><dl><dt>
          
          <code><a href="extract_TOP_coef_samples.html">extract_TOP_coef_samples()</a></code> 
        </dt>
        <dd>Extract alpha and beta coefficients from TOP posterior samples</dd>
      </dl><dl><dt>
          
          <code><a href="extract_TOP_mean_coef.html">extract_TOP_mean_coef()</a></code> 
        </dt>
        <dd>Compute TOP posterior mean coefficients for all three levels of TOP model</dd>
      </dl><dl><dt>
          
          <code><a href="filter_blacklist.html">filter_blacklist()</a></code> 
        </dt>
        <dd>Filter sites in ENCODE blacklist regions</dd>
      </dl><dl><dt>
          
          <code><a href="fit_TOP_M5_model.html">fit_TOP_M5_model()</a></code> 
        </dt>
        <dd>Fit TOP model</dd>
      </dl><dl><dt>
          
          <code><a href="fit_TOP_logistic_M5_model.html">fit_TOP_logistic_M5_model()</a></code> 
        </dt>
        <dd>Fit TOP logistic model</dd>
      </dl><dl><dt>
          
          <code><a href="flank_fimo_sites.html">flank_fimo_sites()</a></code> 
        </dt>
        <dd>Get candidate sites using FIMO motifs with flanking regions</dd>
      </dl><dl><dt>
          
          <code><a href="get_dnase_sites_counts.html">get_dnase_sites_counts()</a></code> 
        </dt>
        <dd>Get DNase count matrices for candidate sites.</dd>
      </dl><dl><dt>
          
          <code><a href="get_total_reads.html">get_total_reads()</a></code> 
        </dt>
        <dd>Get total number of mapped reads from the idxstats file generated by samtools</dd>
      </dl><dl><dt>
          
          <code><a href="load_TOP_samples.html">load_TOP_samples()</a></code> 
        </dt>
        <dd>Load TOP posterior samples</dd>
      </dl><dl><dt>
          
          <code><a href="merge_normalize_dnase_counts.html">merge_normalize_dnase_counts()</a></code> 
        </dt>
        <dd>Merge DNase counts, then normalize merged DNase counts data.</dd>
      </dl><dl><dt>
          
          <code><a href="millipede_binning.html">millipede_binning()</a></code> 
        </dt>
        <dd>Count DNase cuts with different MILLIPEDE binning settings.</dd>
      </dl><dl><dt>
          
          <code><a href="normalize_bin_dnase.html">normalize_bin_dnase()</a></code> 
        </dt>
        <dd>Normalize, binning and transform DNase counts</dd>
      </dl><dl><dt>
          
          <code><a href="normalize_chip.html">normalize_chip()</a></code> 
        </dt>
        <dd>Normalize and transform ChIP counts</dd>
      </dl><dl><dt>
          
          <code><a href="normalize_dnase.html">normalize_dnase()</a></code> 
        </dt>
        <dd>Normalize DNase counts</dd>
      </dl><dl><dt>
          
          <code><a href="predict_TOP.html">predict_TOP()</a></code> 
        </dt>
        <dd>Predict quantitative TF occupancy or TF binding probability</dd>
      </dl><dl><dt>
          
          <code><a href="predict_TOP_logistic_mean_coef.html">predict_TOP_logistic_mean_coef()</a></code> 
        </dt>
        <dd>Predict TF binding by TOP logistic model
with posterior mean of regression coefficients</dd>
      </dl><dl><dt>
          
          <code><a href="predict_TOP_mean_coef.html">predict_TOP_mean_coef()</a></code> 
        </dt>
        <dd>Predict TF occupancy using posterior mean of regression coefficients</dd>
      </dl><dl><dt>
          
          <code><a href="predict_TOP_samples.html">predict_TOP_samples()</a></code> 
        </dt>
        <dd>Predict TF occupancy using posterior samples of regression coefficients</dd>
      </dl><dl><dt>
          
          <code><a href="process_candidate_sites.html">process_candidate_sites()</a></code> 
        </dt>
        <dd>Full process to get candidate sites from FIMO result</dd>
      </dl><dl><dt>
          
          <code><a href="rev_count_bwtool.html">rev_count_bwtool()</a></code> 
        </dt>
        <dd>Flip the counts generated from bwtool for motifs on the reverse (minus) strand</dd>
      </dl><dl><dt>
          
          <code><a href="select_model_coef_mean.html">select_model_coef_mean()</a></code> 
        </dt>
        <dd>Select TOP model hierarchy level</dd>
      </dl><dl><dt>
          
          <code><a href="sort_merge_overlap_bedgraph.html">sort_merge_overlap_bedgraph()</a></code> 
        </dt>
        <dd>Sort and merge the overlapping sites,
remove the overlapping sites with lower occupancy</dd>
      </dl><dl><dt>
          
          <code><a href="track_def_line.html">track_def_line()</a></code> 
        </dt>
        <dd>Set genome browser track parameters for predictions in bedGraph format</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer></div>

  

  

  </body></html>

