<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ TOP</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_preparation.html">Prepare input data</a>
    </li>
    <li>
      <a href="../articles/predict_TF_occupancy_with_trained_model.html">Predict TF occupancy using trained TOP model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_model.html">Train TOP quantitative occupancy model</a>
    </li>
    <li>
      <a href="../articles/train_TOP_logistic_model.html">Train TOP logistic (binary binding) model</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kevinlkx/TOP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_chip_peak_labels_to_sites.html">add_chip_peak_labels_to_sites()</a></code> </p>
        </td>
        <td><p>Add ChIP-seq peak labels to the candidate sites</p></td>
      </tr><tr><td>
          <p><code><a href="add_chip_signals_to_sites.html">add_chip_signals_to_sites()</a></code> </p>
        </td>
        <td><p>Add ChIP-seq signals to the candidate sites</p></td>
      </tr><tr><td>
          <p><code><a href="assemble_partition_training_data.html">assemble_partition_training_data()</a></code> </p>
        </td>
        <td><p>Assemble TOP training data for all TF x cell type combos
in a selected partition</p></td>
      </tr><tr><td>
          <p><code><a href="assemble_training_data.html">assemble_training_data()</a></code> </p>
        </td>
        <td><p>Assemble TOP training data for all TF x cell type combos,
then split training data into 10 partitions</p></td>
      </tr><tr><td>
          <p><code><a href="bin_transform_counts.html">bin_transform_counts()</a></code> </p>
        </td>
        <td><p>Bin and transform count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="combine_TOP_samples.html">combine_TOP_samples()</a></code> </p>
        </td>
        <td><p>Combine and take the average of TOP posterior samples from all partitions</p></td>
      </tr><tr><td>
          <p><code><a href="count_genome_cuts.html">count_genome_cuts()</a></code> </p>
        </td>
        <td><p>Count DNase-seq or ATAC-seq cuts along the genome</p></td>
      </tr><tr><td>
          <p><code><a href="count_normalize_chip.html">count_normalize_chip()</a></code> </p>
        </td>
        <td><p>Count and normalize (and transform) ChIP-seq read coverage</p></td>
      </tr><tr><td>
          <p><code><a href="extract_TOP_coef_samples.html">extract_TOP_coef_samples()</a></code> </p>
        </td>
        <td><p>Extract alpha and beta coefficients from TOP posterior samples</p></td>
      </tr><tr><td>
          <p><code><a href="extract_TOP_mean_coef.html">extract_TOP_mean_coef()</a></code> </p>
        </td>
        <td><p>Extract the posterior mean coefficients for each level of TOP model</p></td>
      </tr><tr><td>
          <p><code><a href="extract_tf_cell_combos.html">extract_tf_cell_combos()</a></code> </p>
        </td>
        <td><p>Extract a table listing the indices and names of TF and cell type
combinations from the assembled training data</p></td>
      </tr><tr><td>
          <p><code><a href="filter_blacklist.html">filter_blacklist()</a></code> </p>
        </td>
        <td><p>Filter candidate sites in blacklist regions</p></td>
      </tr><tr><td>
          <p><code><a href="filter_mapability.html">filter_mapability()</a></code> </p>
        </td>
        <td><p>Filter candidate sites by mapability</p></td>
      </tr><tr><td>
          <p><code><a href="fimo_motif_matches.html">fimo_motif_matches()</a></code> </p>
        </td>
        <td><p>Run FIMO to scan for motif matches</p></td>
      </tr><tr><td>
          <p><code><a href="fit_TOP_M5_model.html">fit_TOP_M5_model()</a></code> </p>
        </td>
        <td><p>Fit TOP model</p></td>
      </tr><tr><td>
          <p><code><a href="fit_TOP_M5_model_jags.html">fit_TOP_M5_model_jags()</a></code> </p>
        </td>
        <td><p>Fit TOP quantitative occupancy model with M5 bins</p></td>
      </tr><tr><td>
          <p><code><a href="fit_TOP_logistic_M5_model_jags.html">fit_TOP_logistic_M5_model_jags()</a></code> </p>
        </td>
        <td><p>Fit TOP logistic model with M5 DNase (or ATAC) bins and binary ChIP labels.</p></td>
      </tr><tr><td>
          <p><code><a href="flank_fimo_sites.html">flank_fimo_sites()</a></code> </p>
        </td>
        <td><p>Get candidate sites using FIMO motifs with flanking regions</p></td>
      </tr><tr><td>
          <p><code><a href="flip_rev_strand_counts.html">flip_rev_strand_counts()</a></code> </p>
        </td>
        <td><p>Flip the counts generated from <code>bwtool</code> for motifs
on the reverse strand</p></td>
      </tr><tr><td>
          <p><code><a href="get_sites_counts.html">get_sites_counts()</a></code> </p>
        </td>
        <td><p>Get count matrices around candidate binding sites</p></td>
      </tr><tr><td>
          <p><code><a href="get_total_reads.html">get_total_reads()</a></code> </p>
        </td>
        <td><p>Get total number of mapped reads from the idxstats file</p></td>
      </tr><tr><td>
          <p><code><a href="index_fa.html">index_fa()</a></code> </p>
        </td>
        <td><p>Index the FASTA file, and generate a 'chrom.sizes' file</p></td>
      </tr><tr><td>
          <p><code><a href="load_TOP_samples.html">load_TOP_samples()</a></code> </p>
        </td>
        <td><p>Load and process TOP posterior samples</p></td>
      </tr><tr><td>
          <p><code><a href="merge_normalize_bin_transform_counts.html">merge_normalize_bin_transform_counts()</a></code> </p>
        </td>
        <td><p>Merge counts from multiple replicates, then normalize,
bin and transform the merged counts</p></td>
      </tr><tr><td>
          <p><code><a href="merge_normalize_counts.html">merge_normalize_counts()</a></code> </p>
        </td>
        <td><p>Merge counts from multiple replicates, then normalize merged counts</p></td>
      </tr><tr><td>
          <p><code><a href="millipede_binning.html">millipede_binning()</a></code> </p>
        </td>
        <td><p>Perform MILLIPEDE binning on count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="normalize_bin_transform_counts.html">normalize_bin_transform_counts()</a></code> </p>
        </td>
        <td><p>Normalize, bin and transform counts</p></td>
      </tr><tr><td>
          <p><code><a href="normalize_chip.html">normalize_chip()</a></code> </p>
        </td>
        <td><p>Normalize (and transform) ChIP-seq read coverage</p></td>
      </tr><tr><td>
          <p><code><a href="normalize_counts.html">normalize_counts()</a></code> </p>
        </td>
        <td><p>Normalize read counts</p></td>
      </tr><tr><td>
          <p><code><a href="predict_TOP.html">predict_TOP()</a></code> </p>
        </td>
        <td><p>Predict quantitative TF occupancy or TF binding probability</p></td>
      </tr><tr><td>
          <p><code><a href="predict_TOP_logistic_mean_coef.html">predict_TOP_logistic_mean_coef()</a></code> </p>
        </td>
        <td><p>Predict TF binding probability by TOP logistic model</p></td>
      </tr><tr><td>
          <p><code><a href="predict_TOP_mean_coef.html">predict_TOP_mean_coef()</a></code> </p>
        </td>
        <td><p>Predict TF occupancy using posterior mean of regression coefficients</p></td>
      </tr><tr><td>
          <p><code><a href="predict_TOP_samples.html">predict_TOP_samples()</a></code> </p>
        </td>
        <td><p>Predict TF occupancy using posterior samples of regression coefficients</p></td>
      </tr><tr><td>
          <p><code><a href="process_candidate_sites.html">process_candidate_sites()</a></code> </p>
        </td>
        <td><p>Obtain and filter candidate sites from FIMO result</p></td>
      </tr><tr><td>
          <p><code><a href="scatterplot.html">scatterplot()</a></code> </p>
        </td>
        <td><p>Simple x-y scatter plot</p></td>
      </tr><tr><td>
          <p><code><a href="scatterplot_predictions.html">scatterplot_predictions()</a></code> </p>
        </td>
        <td><p>Scatter plot of measured and predicted occupancy</p></td>
      </tr><tr><td>
          <p><code><a href="select_features.html">select_features()</a></code> </p>
        </td>
        <td><p>Select PWM and DNase (or ATAC) bin features from the input data frame</p></td>
      </tr><tr><td>
          <p><code><a href="select_model_coef_level.html">select_model_coef_level()</a></code> </p>
        </td>
        <td><p>Select regression coefficients by the TOP hierarchy level</p></td>
      </tr><tr><td>
          <p><code><a href="sort_index_idxstats_bam.html">sort_index_idxstats_bam()</a></code> </p>
        </td>
        <td><p>Sort and index the BAM file, and retrieve the idxstats.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kaixuan Luo, Jianling Zhong, Alex Hartemink.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

