
#' Count and normalize ChIP-seq read coverage
#'
#' @param sites_file Candidate sites file
#' @param chip_bam_files ChIP-seq bam files
#' @param chip_idxstats_files ChIP-seq idxstats files.
#' By default, use the corresponding '.idxstats.txt' files in the same directory
#' of the bam files.
#' @param chrom_size_file Chromosome size file
#' @param ref.size ChIP-Seq reference library size (Default: 10 million)
#' @param transform Transform of ChIP read counts.
#' Options are 'none' (do not transform), 'asinh', 'log2', 'sqrt'.
#' @param bedtools_path Path to bedtools executable
# @importFrom  data.table fread
#' @export
#'
count_normalize_chip = function(sites_file,
                                chip_bam_files,
                                chip_idxstats_files,
                                chrom_size_file,
                                ref.size=1e7,
                                transform = c('none', 'asinh', 'log2', 'sqrt'),
                                bedtools_path='bedtools'){

  if ( Sys.which(bedtools_path) == '' ) {
    stop( 'bedtools could not be executed, set bedtools_path!' )
  }

  if ( length(chip_bam_files) == 0 ) {
    stop('No ChIP-seq files! \n')
  }

  if(missing(chip_idxstats_files)){
    chip_idxstats_files <- paste0(chip_bam_files, '.idxstats.txt')
  }

  # Count total ChIP-seq reads (pool replicates together)
  cat('Counting ChIP-seq read coverage ... \n')
  chip_bam_files <- paste(chip_bam_files, collapse = ' ')
  chip_count_file <- tempfile(pattern = "totalcounts")
  cmd <- paste(bedtools_path, 'coverage -a', sites_file, '-b', chip_bam_files,
               '-counts -sorted -g', chrom_size_file, '>', chip_count_file)
  system(cmd)

  sites_chip.df <- fread(chip_count_file)
  sites <- as.data.frame(fread(sites_file))
  colnames(sites_chip.df) <- c(colnames(sites), 'chip')

  # Normalize (scale) ChIP-seq read counts
  cat('Normalize (scale) ChIP-seq library to', ref.size / 1e6, 'million reads... \n')
  total_readsMapped <- sum(sapply(chip_idxstats_files, get_total_reads, select.chr = TRUE))
  scaling_factor <- ref.size / total_readsMapped
  sites_chip.df$chip <- sites_chip.df$chip * scaling_factor

  # Transform the chip data
  if (transform == 'asinh') {
    sites_chip.df$chip <- asinh(sites_chip.df$chip)
  } else if (transform == 'log2') {
    sites_chip.df$chip <- log2(sites_chip.df$chip + 1)
  } else if (transform == 'sqrt') {
    sites_chip.df$chip <- sqrt(sites_chip.df$chip)
  }

  file.remove(chip_count_file)

  return(sites_chip.df)

}


#' @title Normalize and transform ChIP counts
#'
#' @param chip_counts ChIP-seq counts
#' @param idxstats_file idxstats file generated by samtools
#' @param ref.size ChIP-Seq reference library size (Default: 10 million)
#' @param transform Transform of ChIP read counts.
#' Options are 'none' (do not transform), 'asinh', 'log2', 'sqrt'.
#' @export
#'
normalize_chip <- function(chip_counts,
                           idxstats_file,
                           ref.size=1e7,
                           transform = c('none', 'asinh', 'log2', 'sqrt')) {

  # Count total mapped DNase reads (chr1:22)
  total_readsMapped <- get_total_reads(idxstats_file, select.chr = TRUE)

  # Normalize (scale) ChIP-seq read counts
  cat('Normalize ChIP-seq reads to', ref.size / 1e6, 'million \n')
  scaling_factor <- ref.size / total_readsMapped
  chip_counts <- chip_counts * scaling_factor

  # Transform the ChIP-seq counts
  if (transform == 'asinh') {
    chip_counts <- asinh(chip_counts)
  } else if (transform == 'log2') {
    chip_counts <- log2(chip_counts + 1)
  } else if (transform == 'sqrt') {
    chip_counts <- sqrt(chip_counts)
  }

  return(chip_counts)

}

