
#' @title Get total number of mapped reads from the idxstats file
#' generated by \code{samtools}
#' @description Loads the idxstats file generated by \code{samtools}, and count
#' the total number of mapped reads.
#' @param idxstats_file idxstats file generated by \code{samtools}.
#' @param select.chr Logitic. If TRUE, count reads in
#' chromosomes specified in \code{chrs}.
#' Otherwise, use all chromosomes in the idxstats file.
#' @param chrs Chromosomes to be included (default: chr1, ..., chr22).
#' @return The total number of mapped reads.
#' @export
#'
get_total_reads <- function(idxstats_file, select.chr = TRUE, chrs = paste0('chr', c(1:22))) {
  count.df <- as.data.frame(data.table::fread(idxstats_file, sep = '\t'))
  names(count.df) <- c('chr', 'length', 'reads_mapped', 'reads_unmapped')

  if( select.chr ){
    total_count <- sum(count.df[which(count.df$chr %in% chrs), 'reads_mapped'])
  } else {
    total_count <- sum(count.df[, 'reads_mapped'])
  }

  return(total_count)
}


#' @title Sort and index the BAM file, and retrieve stats (idxstats)
#' @description Uses \code{samtools} to sort and index the BAM file, and
#' retrieve the number of mapped reads using the \code{idxstats} function from
#' \code{samtools}.
#' @param bam_file Input BAM file.
#' @param outdir Output directory (default: save to the directory of the BAM file).
#' @param sort Logical. If TRUE, sort (and index) the BAM file.
#' @param stats Logical. If TRUE, retrieve stats (idxstats) in the index file.
#' @param samtools_path Path to \code{samtools} executable.
#' @export
#'
bam_sort_index_stats <- function(bam_file,
                                 outdir,
                                 sort=TRUE,
                                 stats=TRUE,
                                 samtools_path='samtools') {

  if ( system( paste(samtools_path,'--help'), ignore.stdout=TRUE,ignore.stderr=TRUE ) != 0 ) {
    stop('ERROR: samtools could not be executed. Please install samtools and set samtools_path.')
  }

  if( missing(outdir) ) {
    outdir <- dirname(bam_file)
  }

  if(!dir.exists(outdir)){
    dir.create(outdir, showWarnings = FALSE, recursive = TRUE)
  }

  bam_prefix <- gsub('.bam$', '', basename(bam_file))
  bam_file_sorted <- paste0(outdir, '/', bam_prefix, '.bam')

  if( sort ) {
    # sort and index the bam file
    cat('Sort and index the bam file...\n')
    bam_file_tmp <- paste0(outdir, '/', bam_prefix, '.tmp.bam')
    cmd <- paste(samtools_path, 'sort', bam_file, '-o', bam_file_tmp)
    if(.Platform$OS.type == "windows") shell(cmd) else system(cmd)

    file.rename(from = bam_file_tmp, to = bam_file_sorted)
    cmd <- paste(samtools_path, 'index', bam_file_sorted)
    if(.Platform$OS.type == "windows") shell(cmd) else system(cmd)
  }

  if( stats ){
    # Get index statistics (including the number of mapped reads in the third column)
    cat('Retrieve and print stats in the index file. Result saved to: ', bam_idxstats, '\n', sep='')
    bam_idxstats <- paste0(outdir, '/', bam_prefix, '.bam.idxstats.txt')
    cmd <- paste(samtools_path, 'idxstats', bam_file_sorted, '>', bam_idxstats)
    if(.Platform$OS.type == "windows") shell(cmd) else system(cmd)
  }

}
