
#' @title Index the FASTA file, and generate a 'chrom.sizes' file
#' @description Index the FASTA file, and generate a 'chrom.sizes' file
#' with chromosomes and genome size of each chromosome.
#' @param fa_file FASTA file.
#' @param chromsize_file Path of the output 'chrom.sizes' file.
#' @param outdir Output directory (default: save to the directory of the FASTA file).
#' @importFrom Rsamtools indexFa
#' @export
#'
index_fa <- function(fa_file,
                     chromsize_file='chrom.sizes',
                     outdir=dirname(fa_file)){

  cat('Index', fa_file, 'file and generate chrom.sizes file ...\n')
  faidx_file <- indexFa(fa_file)
  faidx <- read.table(faidx_file, sep='\t')
  write.table(faidx[,1:2], chromsize_file, quote=FALSE,
              col.names=FALSE, row.names=FALSE, sep='\t')

}

#' @title Sort and index the BAM file, and retrieve the idxstats.
#' @description Sort and index the BAM file, and
#' retrieve the idxstats using \code{Rsamtools}.
#' @param bam_file Input BAM file.
#' @param outdir Output directory (default: save to the directory of the BAM file).
#' @param sorted_bam_file Output file name for sorted BAM file if 'sort=TRUE'.
#' @param sort Logical. If TRUE, sort the BAM file.
#' @param index Logical. If TRUE, index the BAM file.
#' @param idxstats Logical. If TRUE, retrieve idxstats from index file.
#' @importFrom Rsamtools sortBam indexBam idxstatsBam
#' @export
#'
sort_index_idxstats_bam <- function(bam_file,
                                    outdir=dirname(bam_file),
                                    sorted_bam_file,
                                    sort=TRUE,
                                    index=TRUE,
                                    idxstats=TRUE) {

  if(!dir.exists(outdir))
    dir.create(outdir)

  if( sort ) {
    # Sort BAM file
    if(missing(sorted_bam_file)){
      out_prefix <- gsub('.bam$','',basename(bam_file))
      sorted_bam_file <- file.path(outdir, paste0(out_prefix, 'sorted.bam'))
    }
    cat('Sort BAM file...\n')
    sortBam(bam_file, sorted_bam_file)
  }else{
    sorted_bam_file <- bam_file
  }

  if( index ) {
    # Index BAM file
    cat('Index BAM file...\n')
    indexBam(sorted_bam_file)
  }

  if( idxstats ){
    # Get index statistics (including the number of mapped reads)
    cat('Retrieve idxstats from the index file...\n')
    out_prefix <- gsub('.bam$','',basename(sorted_bam_file))
    idxstats_file <- file.path(outdir, paste0(out_prefix, '.bam.idxstats.txt'))
    idxstats <- idxstatsBam(sorted_bam_file)
    write.table(idxstats, idxstats_file, col.names=TRUE, row.names = FALSE,
                quote=FALSE, sep='\t')
  }

}


#' @title Get total number of mapped reads from the idxstats file
#' @description Loads the idxstats file, and count
#' the total number of mapped reads generated by \code{bam_sort_index_stats}.
#' @param idxstats_file idxstats file.
#' @param select_chr Logitic. If TRUE, count reads in
#' chromosomes specified in `chrs`.
#' Otherwise, use all chromosomes in the idxstats file.
#' @param chrs Chromosomes to be included (default: chr1, ..., chr22).
#' @return The total number of mapped reads.
#' @export
#'
get_total_reads <- function(idxstats_file, select_chr = TRUE, chrs = paste0('chr', c(1:22))) {
  count.df <- as.data.frame(data.table::fread(idxstats_file, sep = '\t'))
  names(count.df) <- c('chr', 'length', 'reads_mapped', 'reads_unmapped')

  if( select_chr ){
    total_count <- sum(count.df[which(count.df$chr %in% chrs), 'reads_mapped'])
  } else {
    total_count <- sum(count.df[, 'reads_mapped'])
  }

  return(total_count)
}

